<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>pycistarget.motif_enrichment_cistarget &mdash; pycistarget 1.0 documentation</title>
      <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
        <script src="../../_static/jquery.js"></script>
        <script src="../../_static/underscore.js"></script>
        <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../index.html" class="icon icon-home"> pycistarget
          </a>
              <div class="version">
                1.0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tools.html">pycisTarget methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../motif_table.html">SCENIC+ motif collection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../api.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">pycistarget</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../index.html">Module code</a> &raquo;</li>
      <li>pycistarget.motif_enrichment_cistarget</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for pycistarget.motif_enrichment_cistarget</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">typing_extensions</span> <span class="k">import</span> <span class="n">final</span>
<span class="kn">from</span> <span class="nn">ctxcore.genesig</span> <span class="k">import</span> <span class="n">Regulon</span><span class="p">,</span> <span class="n">GeneSignature</span>
<span class="kn">from</span> <span class="nn">ctxcore.recovery</span> <span class="k">import</span> <span class="n">recovery</span><span class="p">,</span> <span class="n">aucs</span> <span class="k">as</span> <span class="n">calc_aucs</span>
<span class="kn">from</span> <span class="nn">ctxcore.recovery</span> <span class="k">import</span> <span class="n">leading_edge4row</span>
<span class="kn">from</span> <span class="nn">ctxcore.rnkdb</span> <span class="k">import</span> <span class="n">FeatherRankingDatabase</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="k">import</span> <span class="n">repeat</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="k">import</span> <span class="n">partial</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">pyranges</span> <span class="k">as</span> <span class="nn">pr</span>
<span class="kn">import</span> <span class="nn">ray</span>
<span class="kn">import</span> <span class="nn">ssl</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="k">import</span> <span class="n">Union</span><span class="p">,</span> <span class="n">Dict</span><span class="p">,</span> <span class="n">Sequence</span><span class="p">,</span> <span class="n">Optional</span>
<span class="kn">import</span> <span class="nn">h5py</span>
<span class="kn">from</span> <span class="nn">collections.abc</span> <span class="k">import</span> <span class="n">Mapping</span>
<span class="kn">from</span> <span class="nn">.utils</span> <span class="k">import</span> <span class="n">is_iterable_not_string</span>

<span class="kn">from</span> <span class="nn">IPython.display</span> <span class="k">import</span> <span class="n">HTML</span>
<span class="n">ssl</span><span class="o">.</span><span class="n">_create_default_https_context</span> <span class="o">=</span> <span class="n">ssl</span><span class="o">.</span><span class="n">_create_unverified_context</span>
<span class="n">pd</span><span class="o">.</span><span class="n">set_option</span><span class="p">(</span><span class="s1">&#39;display.max_colwidth&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>

<span class="c1"># Set stderr to null when using ray.init to avoid ray printing Broken pipe million times</span>
<span class="n">_stderr</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span>
<span class="n">null</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">devnull</span><span class="p">,</span><span class="s1">&#39;wb&#39;</span><span class="p">)</span> 


<span class="kn">from</span> <span class="nn">.utils</span> <span class="k">import</span> <span class="o">*</span>

<div class="viewcode-block" id="cisTargetDatabase"><a class="viewcode-back" href="../../api.html#pycistarget.motif_enrichment_cistarget.cisTargetDatabase">[docs]</a><span class="k">class</span> <span class="nc">cisTargetDatabase</span><span class="p">:</span> 
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    cisTarget Database class.</span>
<span class="sd">    :class:`cisTargetDatabase` contains a dataframe with motifs as rows, regions as columns and rank as</span>
<span class="sd">    values. In addition, is contains a slot to map query regions to regions in the database. For more</span>
<span class="sd">    information on how to generate databases, please visit: https://github.com/aertslab/create_cisTarget_databases</span>
<span class="sd">    </span>
<span class="sd">    Attributes</span>
<span class="sd">    ---------</span>
<span class="sd">    regions_to_db: pd.DataFrame</span>
<span class="sd">        A dataframe containing the mapping between query regions and regions in the database.</span>
<span class="sd">    db_rankings: pd.DataFrame</span>
<span class="sd">        A dataframe with motifs as rows, regions as columns and rank as values.</span>
<span class="sd">    total_regions: int</span>
<span class="sd">        Total number of regions in the database</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> 
                <span class="n">fname</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                <span class="n">region_sets</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">],</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                <span class="n">fraction_overlap</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.4</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize cisTargetDatabase</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ---------</span>
<span class="sd">        fname: str</span>
<span class="sd">            Path to feather file containing the cisTarget database (regions_vs_motifs)</span>
<span class="sd">        region_sets: Dict or pr.PyRanges, optional</span>
<span class="sd">            Dictionary or pr.PyRanges that are going to be analyzed with cistarget. Default: None.</span>
<span class="sd">        name: str, optional</span>
<span class="sd">            Name for the cistarget database. Default: None</span>
<span class="sd">        fraction_overlap: float, optional</span>
<span class="sd">            Minimal overlap between query and regions in the database for the mapping.     </span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">regions_to_db</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">db_rankings</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">total_regions</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">load_db</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span>
                                                          <span class="n">region_sets</span><span class="p">,</span>
                                                          <span class="n">name</span><span class="p">,</span>
                                                          <span class="n">fraction_overlap</span><span class="p">)</span>
<div class="viewcode-block" id="cisTargetDatabase.load_db"><a class="viewcode-back" href="../../api.html#pycistarget.motif_enrichment_cistarget.cisTargetDatabase.load_db">[docs]</a>    <span class="k">def</span> <span class="nf">load_db</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                <span class="n">fname</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                <span class="n">region_sets</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="n">Dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">],</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                <span class="n">fraction_overlap</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.4</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Load cisTargetDatabase</span>
<span class="sd">        </span>
<span class="sd">        Parameters</span>
<span class="sd">        ---------</span>
<span class="sd">        fname: str</span>
<span class="sd">            Path to feather file containing the cisTarget database (regions_vs_motifs)</span>
<span class="sd">        region_sets: Dict or pr.PyRanges, optional</span>
<span class="sd">            Dictionary or pr.PyRanges that are going to be analyzed with cistarget. Default: None.</span>
<span class="sd">        name: str, optional</span>
<span class="sd">            Name for the cistarget database. Default: None</span>
<span class="sd">        fraction_overlap: float, optional</span>
<span class="sd">            Minimal overlap between query and regions in the database for the mapping.     </span>
<span class="sd">            </span>
<span class="sd">        Return</span>
<span class="sd">        ---------</span>
<span class="sd">        target_to_db_dict: pd.DataFrame</span>
<span class="sd">            A dataframe containing the mapping between query regions and regions in the database.</span>
<span class="sd">        db_rankings: pd.DataFrame</span>
<span class="sd">            A dataframe with motifs as rows, regions as columns and rank as values.</span>
<span class="sd">        total_regions: int</span>
<span class="sd">            Total number of regions in the database</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1">#Create logger</span>
        <span class="n">level</span>    <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span>
        <span class="nb">format</span>   <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> </span><span class="si">%(name)-12s</span><span class="s1"> </span><span class="si">%(levelname)-8s</span><span class="s1"> </span><span class="si">%(message)s</span><span class="s1">&#39;</span>
        <span class="n">handlers</span> <span class="o">=</span> <span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(</span><span class="n">stream</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)]</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span> <span class="o">=</span> <span class="n">level</span><span class="p">,</span> <span class="nb">format</span> <span class="o">=</span> <span class="nb">format</span><span class="p">,</span> <span class="n">handlers</span> <span class="o">=</span> <span class="n">handlers</span><span class="p">)</span>
        <span class="n">log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;cisTarget&#39;</span><span class="p">)</span>
        
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Reading cisTarget database&#39;</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">fname</span><span class="p">)</span>
        <span class="n">db</span> <span class="o">=</span> <span class="n">FeatherRankingDatabase</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">)</span>
        <span class="n">total_regions</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">total_genes</span>
        <span class="n">db_regions</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">genes</span>
        <span class="n">prefix</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">if</span> <span class="s1">&#39;__&#39;</span> <span class="ow">in</span> <span class="n">db_regions</span><span class="p">[</span><span class="mi">0</span><span class="p">]:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="n">db_regions</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">db_regions</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">db_regions</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">region_sets</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">region_sets</span><span class="p">)</span> <span class="o">==</span> <span class="nb">dict</span><span class="p">:</span>
                <span class="n">target_to_db_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">x</span><span class="p">:</span> <span class="n">target_to_query</span><span class="p">(</span><span class="n">region_sets</span><span class="p">[</span><span class="n">x</span><span class="p">],</span> <span class="nb">list</span><span class="p">(</span><span class="n">db_regions</span><span class="p">),</span> <span class="n">fraction_overlap</span> <span class="o">=</span> <span class="n">fraction_overlap</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">region_sets</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
                <span class="n">target_regions_in_db</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="nb">sum</span><span class="p">([</span><span class="n">target_to_db_dict</span><span class="p">[</span><span class="n">x</span><span class="p">][</span><span class="s1">&#39;Query&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">target_to_db_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()],[])))</span>
            <span class="k">elif</span> <span class="nb">type</span><span class="p">(</span><span class="n">region_sets</span><span class="p">)</span> <span class="o">==</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">:</span>
                <span class="n">target_to_db</span> <span class="o">=</span> <span class="n">target_to_query</span><span class="p">(</span><span class="n">region_sets</span><span class="p">,</span> <span class="nb">list</span><span class="p">(</span><span class="n">db_regions</span><span class="p">),</span> <span class="n">fraction_overlap</span> <span class="o">=</span> <span class="n">fraction_overlap</span><span class="p">)</span>
                <span class="n">target_to_db</span><span class="o">.</span><span class="n">index</span> <span class="o">=</span> <span class="n">target_to_db</span><span class="p">[</span><span class="s1">&#39;Target&#39;</span><span class="p">]</span>
                <span class="n">target_to_db_dict</span> <span class="o">=</span> <span class="n">target_to_db</span> <span class="c1">#for return purposes</span>
                <span class="n">target_regions_in_db</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">target_to_db</span><span class="p">[</span><span class="s1">&#39;Query&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s1">&#39;region_sets should be either a dict of PyRanges objects or a single PyRanges object, not </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">region_sets</span><span class="p">)))</span>
            <span class="n">name</span><span class="o">=</span><span class="s1">&#39;test&#39;</span>
            <span class="k">if</span> <span class="n">prefix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">target_regions_in_db</span> <span class="o">=</span> <span class="p">[</span><span class="n">prefix</span> <span class="o">+</span> <span class="s1">&#39;__&#39;</span> <span class="o">+</span> <span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">target_regions_in_db</span><span class="p">]</span>
            <span class="n">target_regions_in_db</span> <span class="o">=</span> <span class="n">GeneSignature</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">name</span><span class="p">,</span> <span class="n">gene2weight</span><span class="o">=</span><span class="n">target_regions_in_db</span><span class="p">)</span>
            <span class="n">db_rankings</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">target_regions_in_db</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">prefix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">db_rankings</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;__&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">db_rankings</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s1">&#39;Loading complete cistarget database, this can take a long time and consumes a lot of memory!&#39;</span><span class="p">)</span>
            <span class="n">target_to_db_dict</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">db_rankings</span> <span class="o">=</span> <span class="n">db</span><span class="o">.</span><span class="n">load_full</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">target_to_db_dict</span><span class="p">,</span> <span class="n">db_rankings</span><span class="p">,</span> <span class="n">total_regions</span></div></div>


<span class="c1"># cisTarget class</span>
<div class="viewcode-block" id="cisTarget"><a class="viewcode-back" href="../../api.html#pycistarget.motif_enrichment_cistarget.cisTarget">[docs]</a><span class="k">class</span> <span class="nc">cisTarget</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    cisTarget class.</span>
<span class="sd">    :class:`cisTarget` contains method for motif enrichment analysis on sets of regions. </span>
<span class="sd">    </span>
<span class="sd">    Attributes</span>
<span class="sd">    ---------</span>
<span class="sd">    regions_to_db: pd.DataFrame</span>
<span class="sd">        A dataframe containing the mapping between query regions and regions in the database.</span>
<span class="sd">    region_set: pr.PyRanges</span>
<span class="sd">        A PyRanges containing region coordinates for the regions to be analyzed.</span>
<span class="sd">    name: str</span>
<span class="sd">        Analysis name</span>
<span class="sd">    specie: str</span>
<span class="sd">        Specie from which genomic coordinates come from</span>
<span class="sd">    auc_threshold: float, optional</span>
<span class="sd">        The fraction of the ranked genome to take into account for the calculation of the</span>
<span class="sd">        Area Under the recovery Curve. Default: 0.005</span>
<span class="sd">    nes_threshold: float, optional</span>
<span class="sd">        The Normalized Enrichment Score (NES) threshold to select enriched features. Default: 3.0</span>
<span class="sd">    rank_threshold: float, optional</span>
<span class="sd">        The total number of ranked genes to take into account when creating a recovery curve.</span>
<span class="sd">        Default: 0.05</span>
<span class="sd">    path_to_motif_annotations: str, optional</span>
<span class="sd">        Path to motif annotations. If not provided, they will be downloaded from </span>
<span class="sd">        https://resources.aertslab.org based on the specie name provided (only possible for mus_musculus,</span>
<span class="sd">        homo_sapiens and drosophila_melanogaster). Default: None</span>
<span class="sd">    annotation_version: str, optional</span>
<span class="sd">        Motif collection version. Default: v9</span>
<span class="sd">    annotation: List, optional</span>
<span class="sd">        Annotation to use for forming cistromes. It can be &#39;Direct_annot&#39; (direct evidence that the motif is </span>
<span class="sd">        linked to that TF), &#39;Motif_similarity_annot&#39; (based on tomtom motif similarity), &#39;Orthology_annot&#39;</span>
<span class="sd">        (based on orthology with a TF that is directly linked to that motif) or &#39;Motif_similarity_and_Orthology_annot&#39;.</span>
<span class="sd">        Default: [&#39;Direct_annot&#39;, &#39;Motif_similarity_annot&#39;, &#39;Orthology_annot&#39;, &#39;Motif_similarity_and_Orthology_annot&#39;]</span>
<span class="sd">    motif_similarity_fdr: float, optional</span>
<span class="sd">        Minimal motif similarity value to consider two motifs similar. Default: 0.001</span>
<span class="sd">    orthologous_identity_threshold: float, optional</span>
<span class="sd">        Minimal orthology value for considering two TFs orthologous. Default: 0.0</span>
<span class="sd">    motifs_to_use: List, optional</span>
<span class="sd">        A subset of motifs to use for the analysis. Default: None (All)</span>
<span class="sd">    motif_enrichment: pd.DataFrame</span>
<span class="sd">        A dataframe containing motif enrichment results</span>
<span class="sd">    motif_hits: Dict</span>
<span class="sd">        A dictionary containing regions that are considered enriched for each motif.</span>
<span class="sd">    cistromes: Dict</span>
<span class="sd">        A dictionary containing TF cistromes. Cistromes with no extension contain regions linked to directly</span>
<span class="sd">        annotated motifs, while &#39;_extended&#39; cistromes can contain regions linked to motifs annotated by </span>
<span class="sd">        similarity or orthology.</span>
<span class="sd">        </span>
<span class="sd">    References</span>
<span class="sd">    ---------</span>
<span class="sd">    Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.</span>
<span class="sd">    Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">def</span> <span class="nf">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> 
                 <span class="n">region_set</span><span class="p">:</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">,</span>
                 <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                 <span class="n">specie</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                 <span class="n">auc_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.005</span><span class="p">,</span>
                 <span class="n">nes_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
                 <span class="n">rank_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
                 <span class="n">path_to_motif_annotations</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                 <span class="n">annotation_version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;v9&#39;</span><span class="p">,</span>
                 <span class="n">annotation</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Direct_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_similarity_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Orthology_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_similarity_and_Orthology_annot&#39;</span><span class="p">],</span>
                 <span class="n">motif_similarity_fdr</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
                 <span class="n">orthologous_identity_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
                 <span class="n">motifs_to_use</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Initialize cisTarget class.</span>

<span class="sd">        Parameters</span>
<span class="sd">        ---------</span>
<span class="sd">        ctx_db: :class:`cisTargetDatabase`</span>
<span class="sd">            A cistarget database object.</span>
<span class="sd">        region_set: pr.PyRanges</span>
<span class="sd">            A PyRanges containing region coordinates for the regions to be analyzed.</span>
<span class="sd">        name: str</span>
<span class="sd">            Analysis name</span>
<span class="sd">        specie: str</span>
<span class="sd">            Specie from which genomic coordinates come from</span>
<span class="sd">        auc_threshold: float, optional</span>
<span class="sd">            The fraction of the ranked genome to take into account for the calculation of the</span>
<span class="sd">            Area Under the recovery Curve. Default: 0.005</span>
<span class="sd">        nes_threshold: float, optional</span>
<span class="sd">            The Normalized Enrichment Score (NES) threshold to select enriched features. Default: 3.0</span>
<span class="sd">        rank_threshold: float, optional</span>
<span class="sd">            The total number of ranked genes to take into account when creating a recovery curve.</span>
<span class="sd">            Default: 0.05</span>
<span class="sd">        path_to_motif_annotations: str, optional</span>
<span class="sd">            Path to motif annotations. If not provided, they will be downloaded from </span>
<span class="sd">            https://resources.aertslab.org based on the specie name provided (only possible for mus_musculus,</span>
<span class="sd">            homo_sapiens and drosophila_melanogaster). Default: None</span>
<span class="sd">        annotation_version: str, optional</span>
<span class="sd">            Motif collection version. Default: v9</span>
<span class="sd">        annotation: List, optional</span>
<span class="sd">            Annotation to use for forming cistromes. It can be &#39;Direct_annot&#39; (direct evidence that the motif is </span>
<span class="sd">            linked to that TF), &#39;Motif_similarity_annot&#39; (based on tomtom motif similarity), &#39;Orthology_annot&#39;</span>
<span class="sd">            (based on orthology with a TF that is directly linked to that motif) or &#39;Motif_similarity_and_Orthology_annot&#39;.</span>
<span class="sd">            Default: [&#39;Direct_annot&#39;, &#39;Motif_similarity_annot&#39;, &#39;Orthology_annot&#39;, &#39;Motif_similarity_and_Orthology_annot&#39;]</span>
<span class="sd">        motif_similarity_fdr: float, optional</span>
<span class="sd">            Minimal motif similarity value to consider two motifs similar. Default: 0.001</span>
<span class="sd">        orthologous_identity_threshold: float, optional</span>
<span class="sd">            Minimal orthology value for considering two TFs orthologous. Default: 0.0</span>
<span class="sd">        motifs_to_use: List, optional</span>
<span class="sd">            A subset of motifs to use for the analysis. Default: None (All)</span>
<span class="sd">    </span>
<span class="sd">        References</span>
<span class="sd">        ---------</span>
<span class="sd">        Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.</span>
<span class="sd">        Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">region_set</span> <span class="o">=</span> <span class="n">region_set</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">specie</span> <span class="o">=</span> <span class="n">specie</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">auc_threshold</span> <span class="o">=</span> <span class="n">auc_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">nes_threshold</span> <span class="o">=</span> <span class="n">nes_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">rank_threshold</span> <span class="o">=</span> <span class="n">rank_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotation_version</span> <span class="o">=</span> <span class="n">annotation_version</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span> <span class="o">=</span> <span class="n">annotation</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">path_to_motif_annotations</span> <span class="o">=</span> <span class="n">path_to_motif_annotations</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_similarity_fdr</span> <span class="o">=</span> <span class="n">motif_similarity_fdr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">orthologous_identity_threshold</span> <span class="o">=</span> <span class="n">orthologous_identity_threshold</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motifs_to_use</span> <span class="o">=</span> <span class="n">motifs_to_use</span>
        
<div class="viewcode-block" id="cisTarget.run_ctx"><a class="viewcode-back" href="../../api.html#pycistarget.motif_enrichment_cistarget.cisTarget.run_ctx">[docs]</a>    <span class="k">def</span> <span class="nf">run_ctx</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
            <span class="n">ctx_db</span><span class="p">:</span> <span class="n">cisTargetDatabase</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Run cisTarget</span>

<span class="sd">        Parameters</span>
<span class="sd">        ---------</span>
<span class="sd">        ctx_db: :class:`cisTargetDatabase`</span>
<span class="sd">            A cistarget database object.</span>
<span class="sd">    </span>
<span class="sd">        References</span>
<span class="sd">        ---------</span>
<span class="sd">        Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.</span>
<span class="sd">        Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</span>
<span class="sd">        &quot;&quot;&quot;</span>
        
        <span class="c1"># Create logger</span>
        <span class="n">level</span>    <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span>
        <span class="nb">format</span>   <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> </span><span class="si">%(name)-12s</span><span class="s1"> </span><span class="si">%(levelname)-8s</span><span class="s1"> </span><span class="si">%(message)s</span><span class="s1">&#39;</span>
        <span class="n">handlers</span> <span class="o">=</span> <span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(</span><span class="n">stream</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)]</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span> <span class="o">=</span> <span class="n">level</span><span class="p">,</span> <span class="nb">format</span> <span class="o">=</span> <span class="nb">format</span><span class="p">,</span> <span class="n">handlers</span> <span class="o">=</span> <span class="n">handlers</span><span class="p">)</span>
        <span class="n">log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;cisTarget&#39;</span><span class="p">)</span>

        <span class="c1">#Hardcoded values</span>
        <span class="n">COLUMN_NAME_NES</span> <span class="o">=</span> <span class="s2">&quot;NES&quot;</span>
        <span class="n">COLUMN_NAME_AUC</span> <span class="o">=</span> <span class="s2">&quot;AUC&quot;</span>
        <span class="n">COLUMN_NAME_GRP</span> <span class="o">=</span> <span class="s2">&quot;GROUP&quot;</span>
        <span class="n">COLUMN_NAME_MOTIF_ID</span> <span class="o">=</span> <span class="s2">&quot;MotifID&quot;</span>
        <span class="n">COLUMN_NAME_TARGET_GENES</span> <span class="o">=</span> <span class="s2">&quot;TargetRegions&quot;</span>
        <span class="n">COLUMN_NAME_RANK_AT_MAX</span> <span class="o">=</span> <span class="s2">&quot;RankAtMax&quot;</span>
        <span class="n">COLUMN_NAME_TYPE</span> <span class="o">=</span> <span class="s2">&quot;Type&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">regions_to_db</span> <span class="o">=</span> <span class="n">ctx_db</span><span class="o">.</span><span class="n">regions_to_db</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="k">if</span> <span class="nb">type</span><span class="p">(</span><span class="n">ctx_db</span><span class="o">.</span><span class="n">regions_to_db</span><span class="p">)</span> <span class="o">==</span> <span class="nb">dict</span> <span class="k">else</span> <span class="n">ctx_db</span><span class="o">.</span><span class="n">regions_to_db</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">coord_to_region_names</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">region_set</span><span class="p">))</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">ctx_db</span><span class="o">.</span><span class="n">regions_to_db</span><span class="p">[</span><span class="s1">&#39;Target&#39;</span><span class="p">])]</span>

        <span class="c1"># Log</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Running cisTarget for </span><span class="si">{}</span><span class="s2"> which has </span><span class="si">{}</span><span class="s2"> regions&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">regions_to_db</span><span class="p">[</span><span class="s1">&#39;Query&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())))</span>
        <span class="c1"># Load signature as Regulon</span>
        <span class="n">region_set_signature</span> <span class="o">=</span> <span class="n">region_sets_to_signature</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">regions_to_db</span><span class="p">[</span><span class="s1">&#39;Query&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),</span> <span class="n">region_set_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="c1"># Get regions</span>
        <span class="n">regions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">region_set_signature</span><span class="o">.</span><span class="n">genes</span><span class="p">))</span>
        
        <span class="c1">#subset rankings database on motifs and regions</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">motifs_to_use</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Using only user provided motifs&#39;</span><span class="p">)</span>
            <span class="n">motifs_not_in_db</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">difference</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs_to_use</span><span class="p">),</span> <span class="nb">set</span><span class="p">(</span><span class="n">ctx_db</span><span class="o">.</span><span class="n">db_rankings</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">))</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">motifs_not_in_db</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Some motifs provided by the parameter &lt;motifs_to_use&gt; are not in the rankings database: </span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">motifs_not_in_db</span><span class="p">))</span>
            <span class="n">motifs_to_use</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motifs_to_use</span><span class="p">)</span> <span class="o">&amp;</span> <span class="nb">set</span><span class="p">(</span><span class="n">ctx_db</span><span class="o">.</span><span class="n">db_rankings</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
            <span class="n">db_rankings_regions</span> <span class="o">=</span> <span class="n">ctx_db</span><span class="o">.</span><span class="n">db_rankings</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">motifs_to_use</span><span class="p">,</span> <span class="n">regions</span><span class="p">]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">db_rankings_regions</span> <span class="o">=</span> <span class="n">ctx_db</span><span class="o">.</span><span class="n">db_rankings</span><span class="p">[</span><span class="n">regions</span><span class="p">]</span>

        <span class="c1">#Get features, rankings and weights</span>
        <span class="n">features</span><span class="p">,</span> <span class="n">rankings</span> <span class="o">=</span> <span class="n">db_rankings_regions</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span><span class="p">,</span> <span class="n">db_rankings_regions</span><span class="o">.</span><span class="n">values</span>
        <span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">regions</span><span class="p">)))</span>
        <span class="c1"># Calculate recovery curves, AUC and NES values.</span>
        <span class="n">aucs</span> <span class="o">=</span> <span class="n">calc_aucs</span><span class="p">(</span><span class="n">db_rankings_regions</span><span class="p">,</span> <span class="n">ctx_db</span><span class="o">.</span><span class="n">total_regions</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">auc_threshold</span><span class="p">)</span>
        <span class="n">ness</span> <span class="o">=</span> <span class="p">(</span><span class="n">aucs</span> <span class="o">-</span> <span class="n">aucs</span><span class="o">.</span><span class="n">mean</span><span class="p">())</span> <span class="o">/</span> <span class="n">aucs</span><span class="o">.</span><span class="n">std</span><span class="p">()</span>
        <span class="c1"># Keep only features that are enriched, i.e. NES sufficiently high.</span>
        <span class="n">enriched_features_idx</span> <span class="o">=</span> <span class="n">ness</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">nes_threshold</span>
        <span class="c1"># Make dataframe</span>
        <span class="n">enriched_features</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">Index</span><span class="p">(</span><span class="n">features</span><span class="p">[</span><span class="n">enriched_features_idx</span><span class="p">],</span> <span class="n">name</span> <span class="o">=</span> <span class="n">COLUMN_NAME_MOTIF_ID</span><span class="p">),</span>
                                    <span class="n">data</span><span class="o">=</span><span class="p">{</span><span class="n">COLUMN_NAME_NES</span><span class="p">:</span> <span class="n">ness</span><span class="p">[</span><span class="n">enriched_features_idx</span><span class="p">],</span>
                                        <span class="n">COLUMN_NAME_AUC</span><span class="p">:</span> <span class="n">aucs</span><span class="p">[</span><span class="n">enriched_features_idx</span><span class="p">],</span>
                                        <span class="n">COLUMN_NAME_GRP</span><span class="p">:</span> <span class="n">repeat</span><span class="p">(</span><span class="n">region_set_signature</span><span class="o">.</span><span class="n">transcription_factor</span><span class="p">,</span> <span class="nb">sum</span><span class="p">(</span><span class="n">enriched_features_idx</span><span class="p">))})</span>
        <span class="c1"># Recovery analysis</span>
        <span class="n">rccs</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">recovery</span><span class="p">(</span><span class="n">db_rankings_regions</span><span class="p">,</span> <span class="n">ctx_db</span><span class="o">.</span><span class="n">total_regions</span><span class="p">,</span> <span class="n">weights</span><span class="p">,</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rank_threshold</span><span class="o">*</span><span class="n">ctx_db</span><span class="o">.</span><span class="n">total_regions</span><span class="p">),</span> <span class="bp">self</span><span class="o">.</span><span class="n">auc_threshold</span><span class="p">,</span> <span class="n">no_auc</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>  
        <span class="n">avgrcc</span> <span class="o">=</span> <span class="n">rccs</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>        
        <span class="n">avg2stdrcc</span> <span class="o">=</span> <span class="n">avgrcc</span> <span class="o">+</span> <span class="mf">2.0</span> <span class="o">*</span> <span class="n">rccs</span><span class="o">.</span><span class="n">std</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        <span class="c1"># Select features</span>
        <span class="n">rccs</span> <span class="o">=</span> <span class="n">rccs</span><span class="p">[</span><span class="n">enriched_features_idx</span><span class="p">,</span> <span class="p">:]</span>
        <span class="n">rankings</span> <span class="o">=</span> <span class="n">rankings</span><span class="p">[</span><span class="n">enriched_features_idx</span><span class="p">,</span> <span class="p">:]</span>
        <span class="c1"># Format df</span>
        <span class="n">enriched_features</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_tuples</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">repeat</span><span class="p">(</span><span class="s2">&quot;Enrichment&quot;</span><span class="p">),</span>
                                                                        <span class="n">enriched_features</span><span class="o">.</span><span class="n">columns</span><span class="p">)))</span>
        <span class="n">df_rnks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">enriched_features</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
                            <span class="n">columns</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_tuples</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">repeat</span><span class="p">(</span><span class="s2">&quot;Ranking&quot;</span><span class="p">),</span> <span class="n">regions</span><span class="p">))),</span>
                            <span class="n">data</span><span class="o">=</span><span class="n">rankings</span><span class="p">)</span>
        <span class="n">df_rccs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">index</span><span class="o">=</span><span class="n">enriched_features</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
                            <span class="n">columns</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">MultiIndex</span><span class="o">.</span><span class="n">from_tuples</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="nb">zip</span><span class="p">(</span><span class="n">repeat</span><span class="p">(</span><span class="s2">&quot;Recovery&quot;</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">rank_threshold</span><span class="o">*</span><span class="n">ctx_db</span><span class="o">.</span><span class="n">total_regions</span><span class="p">))))),</span>
                            <span class="n">data</span><span class="o">=</span><span class="n">rccs</span><span class="p">)</span>
        <span class="n">enriched_features</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">enriched_features</span><span class="p">,</span> <span class="n">df_rccs</span><span class="p">,</span> <span class="n">df_rnks</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="c1"># Calculate the leading edges for each row. Always return importance from gene inference phase.</span>
        <span class="n">weights</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">asarray</span><span class="p">([</span><span class="n">region_set_signature</span><span class="p">[</span><span class="n">region</span><span class="p">]</span> <span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">regions</span><span class="p">])</span>
        <span class="n">enriched_features</span><span class="p">[[(</span><span class="s2">&quot;Enrichment&quot;</span><span class="p">,</span> <span class="n">COLUMN_NAME_TARGET_GENES</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;Enrichment&quot;</span><span class="p">,</span> <span class="n">COLUMN_NAME_RANK_AT_MAX</span><span class="p">)]]</span> <span class="o">=</span> <span class="n">enriched_features</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
            <span class="n">partial</span><span class="p">(</span><span class="n">leading_edge4row</span><span class="p">,</span> <span class="n">avg2stdrcc</span><span class="o">=</span><span class="n">avg2stdrcc</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="n">regions</span><span class="p">,</span> <span class="n">weights</span><span class="o">=</span><span class="n">weights</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">enriched_features</span> <span class="o">=</span> <span class="n">enriched_features</span><span class="p">[</span><span class="s1">&#39;Enrichment&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">rename_axis</span><span class="p">(</span><span class="kc">None</span><span class="p">)</span>
        <span class="c1"># Format enriched features</span>
        <span class="n">enriched_features</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="s1">&#39;AUC&#39;</span><span class="p">,</span> <span class="s1">&#39;Region_set&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_hits&#39;</span><span class="p">,</span> <span class="s1">&#39;Rank_at_max&#39;</span><span class="p">]</span>
        <span class="n">enriched_features</span> <span class="o">=</span> <span class="n">enriched_features</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_enrichment</span> <span class="o">=</span> <span class="n">enriched_features</span><span class="p">[[</span><span class="s1">&#39;Region_set&#39;</span><span class="p">,</span> <span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="s1">&#39;AUC&#39;</span><span class="p">,</span> <span class="s1">&#39;Rank_at_max&#39;</span><span class="p">]]</span>
        <span class="c1"># Annotation</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Annotating motifs for &quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">add_motif_annotation_cistarget</span><span class="p">()</span>
        <span class="c1"># Motif hits</span>
        <span class="n">db_motif_hits</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="p">[</span><span class="n">enriched_features</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">key</span><span class="p">,</span> <span class="s1">&#39;Motif_hits&#39;</span><span class="p">][</span><span class="n">i</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">enriched_features</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">key</span><span class="p">,</span> <span class="s1">&#39;Motif_hits&#39;</span><span class="p">]))]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">enriched_features</span><span class="o">.</span><span class="n">index</span><span class="p">}</span>
        <span class="n">rs_motif_hits</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">regions_to_db</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">regions_to_db</span><span class="p">[</span><span class="s1">&#39;Query&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">db_motif_hits</span><span class="p">[</span><span class="n">key</span><span class="p">]),</span> <span class="s1">&#39;Target&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">db_motif_hits</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_hits</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Database&#39;</span><span class="p">:</span> <span class="n">db_motif_hits</span><span class="p">,</span> <span class="s1">&#39;Region_set&#39;</span><span class="p">:</span> <span class="n">rs_motif_hits</span><span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_enrichment</span><span class="p">[</span><span class="s1">&#39;Motif_hits&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">db_motif_hits</span><span class="p">[</span><span class="n">i</span><span class="p">])</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">db_motif_hits</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
        <span class="c1"># Cistromes</span>
        <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Getting cistromes for &quot;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">cistromes_db</span> <span class="o">=</span> <span class="n">get_cistromes_per_region_set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_enrichment</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_hits</span><span class="p">[</span><span class="s1">&#39;Database&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">)</span>
        <span class="n">cistromes_rs</span> <span class="o">=</span> <span class="n">get_cistromes_per_region_set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_enrichment</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_hits</span><span class="p">[</span><span class="s1">&#39;Region_set&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">cistromes</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Database&#39;</span><span class="p">:</span> <span class="n">cistromes_db</span><span class="p">,</span> <span class="s1">&#39;Region_set&#39;</span><span class="p">:</span> <span class="n">cistromes_rs</span><span class="p">}</span></div>
        
<div class="viewcode-block" id="cisTarget.add_motif_annotation_cistarget"><a class="viewcode-back" href="../../api.html#pycistarget.motif_enrichment_cistarget.cisTarget.add_motif_annotation_cistarget">[docs]</a>    <span class="k">def</span> <span class="nf">add_motif_annotation_cistarget</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span>
                       <span class="n">add_logo</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">bool</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Add motif annotation</span>

<span class="sd">        Parameters</span>
<span class="sd">        ---------</span>
<span class="sd">        add_logo: boolean, optional</span>
<span class="sd">            Whether to add the motif logo to the motif enrichment dataframe</span>
<span class="sd">    </span>
<span class="sd">        References</span>
<span class="sd">        ---------</span>
<span class="sd">        Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.</span>
<span class="sd">        Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Create cisTarget logger</span>
        <span class="n">level</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span>
        <span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> </span><span class="si">%(name)-12s</span><span class="s1"> </span><span class="si">%(levelname)-8s</span><span class="s1"> </span><span class="si">%(message)s</span><span class="s1">&#39;</span>
        <span class="n">handlers</span> <span class="o">=</span> <span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(</span><span class="n">stream</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)]</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">level</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="nb">format</span><span class="p">,</span> <span class="n">handlers</span><span class="o">=</span><span class="n">handlers</span><span class="p">)</span>
        <span class="n">log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;cisTarget&#39;</span><span class="p">)</span>

        <span class="c1"># Read motif annotation. </span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">annot_df</span> <span class="o">=</span> <span class="n">load_motif_annotations</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">specie</span><span class="p">,</span>
                                          <span class="n">version</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation_version</span><span class="p">,</span>
                                          <span class="n">fname</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">path_to_motif_annotations</span><span class="p">,</span>
                                          <span class="n">motif_similarity_fdr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_similarity_fdr</span><span class="p">,</span>
                                          <span class="n">orthologous_identity_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">orthologous_identity_threshold</span><span class="p">)</span>
            <span class="n">motif_enrichment_w_annot</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_enrichment</span><span class="p">,</span> <span class="n">annot_df</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">motif_enrichment</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">(),:]</span>
        <span class="k">except</span><span class="p">:</span>
            <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Unable to load annotation for &#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">specie</span><span class="p">)</span>
            <span class="n">annot_df</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="n">motif_enrichment_w_annot</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">motif_enrichment</span>
        <span class="c1"># Add info to elements in dict</span>
        <span class="k">if</span> <span class="n">add_logo</span> <span class="o">==</span> <span class="kc">True</span><span class="p">:</span>
            <span class="n">motif_enrichment_w_annot</span><span class="p">[</span><span class="s1">&#39;Logo&#39;</span><span class="p">]</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;&lt;img src=&quot;&#39;</span> <span class="o">+</span><span class="s1">&#39;https://motifcollections.aertslab.org/&#39;</span> <span class="o">+</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation_version</span> <span class="o">+</span> <span class="s1">&#39;/logos/&#39;</span><span class="o">+</span> <span class="n">motif_enrichment_w_annot</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[</span><span class="n">i</span><span class="p">]</span> <span class="o">+</span> <span class="s1">&#39;.png&#39;</span> <span class="o">+</span> <span class="s1">&#39;&quot; width=&quot;200&quot; &gt;&#39;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">motif_enrichment_w_annot</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])]</span>
            <span class="k">if</span> <span class="n">annot_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">motif_enrichment_w_annot</span> <span class="o">=</span> <span class="n">motif_enrichment_w_annot</span><span class="p">[</span><span class="nb">sum</span><span class="p">([[</span><span class="s1">&#39;Logo&#39;</span><span class="p">,</span> <span class="s1">&#39;Region_set&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="s1">&#39;AUC&#39;</span><span class="p">,</span> <span class="s1">&#39;Rank_at_max&#39;</span><span class="p">]],</span> <span class="p">[])]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">motif_enrichment_w_annot</span> <span class="o">=</span> <span class="n">motif_enrichment_w_annot</span><span class="p">[[</span><span class="s1">&#39;Logo&#39;</span><span class="p">,</span> <span class="s1">&#39;Region_set&#39;</span><span class="p">,</span> <span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="s1">&#39;AUC&#39;</span><span class="p">,</span> <span class="s1">&#39;Rank_at_max&#39;</span><span class="p">]]</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">annot_df</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">motif_enrichment_w_annot</span> <span class="o">=</span> <span class="n">motif_enrichment_w_annot</span><span class="p">[</span><span class="nb">sum</span><span class="p">([[</span><span class="s1">&#39;Region_set&#39;</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">annotation</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="s1">&#39;AUC&#39;</span><span class="p">,</span> <span class="s1">&#39;Rank_at_max&#39;</span><span class="p">]],</span> <span class="p">[])]</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">motif_enrichment_w_annot</span> <span class="o">=</span> <span class="n">motif_enrichment_w_annot</span><span class="p">[[</span><span class="s1">&#39;Region_set&#39;</span><span class="p">,</span> <span class="s1">&#39;NES&#39;</span><span class="p">,</span> <span class="s1">&#39;AUC&#39;</span><span class="p">,</span> <span class="s1">&#39;Rank_at_max&#39;</span><span class="p">]]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">motif_enrichment</span> <span class="o">=</span> <span class="n">motif_enrichment_w_annot</span> </div></div>


<span class="c1"># Run cisTarget            </span>
<div class="viewcode-block" id="run_cistarget"><a class="viewcode-back" href="../../api.html#pycistarget.motif_enrichment_cistarget.run_cistarget">[docs]</a><span class="k">def</span> <span class="nf">run_cistarget</span><span class="p">(</span><span class="n">ctx_db</span><span class="p">:</span> <span class="n">cisTargetDatabase</span><span class="p">,</span>
                               <span class="n">region_sets</span><span class="p">:</span> <span class="nb">dict</span><span class="p">,</span>
                               <span class="n">specie</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
                               <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;cisTarget&#39;</span><span class="p">,</span>
                               <span class="n">fraction_overlap</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.4</span><span class="p">,</span>
                               <span class="n">auc_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.005</span><span class="p">,</span>
                               <span class="n">nes_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
                               <span class="n">rank_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
                               <span class="n">path_to_motif_annotations</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                               <span class="n">annotation_version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;v9&#39;</span><span class="p">,</span>
                               <span class="n">annotation</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Direct_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_similarity_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Orthology_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_similarity_and_Orthology_annot&#39;</span><span class="p">],</span>
                               <span class="n">motif_similarity_fdr</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
                               <span class="n">orthologous_identity_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
                               <span class="n">n_cpu</span> <span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                               <span class="n">motifs_to_use</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
                               <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run cisTarget.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ---------</span>
<span class="sd">    ctx_db: :class:`cisTargetDatabase`</span>
<span class="sd">        A cistarget database object.</span>
<span class="sd">    region_sets: Dict</span>
<span class="sd">        A dictionary of PyRanges containing region coordinates for the region sets to be analyzed.</span>
<span class="sd">    specie: str</span>
<span class="sd">        Specie from which genomic coordinates come from</span>
<span class="sd">    name: str, optional</span>
<span class="sd">        Analysis name. Default: cisTarget</span>
<span class="sd">    fraction_overlap: float, optional</span>
<span class="sd">        Minimal overlap between query and regions in the database for the mapping. Default: 0.4</span>
<span class="sd">    auc_threshold: float, optional</span>
<span class="sd">        The fraction of the ranked genome to take into account for the calculation of the</span>
<span class="sd">        Area Under the recovery Curve. Default: 0.005</span>
<span class="sd">    nes_threshold: float, optional</span>
<span class="sd">        The Normalized Enrichment Score (NES) threshold to select enriched features. Default: 3.0</span>
<span class="sd">    rank_threshold: float, optional</span>
<span class="sd">        The total number of ranked genes to take into account when creating a recovery curve.</span>
<span class="sd">        Default: 0.05</span>
<span class="sd">    path_to_motif_annotations: str, optional</span>
<span class="sd">        Path to motif annotations. If not provided, they will be downloaded from </span>
<span class="sd">        https://resources.aertslab.org based on the specie name provided (only possible for mus_musculus,</span>
<span class="sd">        homo_sapiens and drosophila_melanogaster). Default: None</span>
<span class="sd">    annotation_version: str, optional</span>
<span class="sd">        Motif collection version. Default: v9</span>
<span class="sd">    annotation: List, optional</span>
<span class="sd">        Annotation to use for forming cistromes. It can be &#39;Direct_annot&#39; (direct evidence that the motif is </span>
<span class="sd">        linked to that TF), &#39;Motif_similarity_annot&#39; (based on tomtom motif similarity), &#39;Orthology_annot&#39;</span>
<span class="sd">        (based on orthology with a TF that is directly linked to that motif) or &#39;Motif_similarity_and_Orthology_annot&#39;.</span>
<span class="sd">        Default: [&#39;Direct_annot&#39;, &#39;Motif_similarity_annot&#39;, &#39;Orthology_annot&#39;, &#39;Motif_similarity_and_Orthology_annot&#39;]</span>
<span class="sd">    motif_similarity_fdr: float, optional</span>
<span class="sd">        Minimal motif similarity value to consider two motifs similar. Default: 0.001</span>
<span class="sd">    orthologous_identity_threshold: float, optional</span>
<span class="sd">        Minimal orthology value for considering two TFs orthologous. Default: 0.0</span>
<span class="sd">    n_cpu: int, optional</span>
<span class="sd">        Number of cores to use. If 1, ray will not be used. Default: 1</span>
<span class="sd">    motifs_to_use: List, optional</span>
<span class="sd">        A subset of motifs to use for the analysis. Default: None (All)</span>
<span class="sd">    **kwargs:</span>
<span class="sd">        Extra parameters to pass to `ray.init()`</span>
<span class="sd">        </span>
<span class="sd">    Return</span>
<span class="sd">    ---------</span>
<span class="sd">        A dictionary with one :class:`cisTarget` object per analysis.</span>
<span class="sd">        </span>
<span class="sd">    References</span>
<span class="sd">    ---------</span>
<span class="sd">    Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.</span>
<span class="sd">    Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Create cisTarget logger</span>
    <span class="n">level</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">INFO</span>
    <span class="nb">format</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="si">%(asctime)s</span><span class="s1"> </span><span class="si">%(name)-12s</span><span class="s1"> </span><span class="si">%(levelname)-8s</span><span class="s1"> </span><span class="si">%(message)s</span><span class="s1">&#39;</span>
    <span class="n">handlers</span> <span class="o">=</span> <span class="p">[</span><span class="n">logging</span><span class="o">.</span><span class="n">StreamHandler</span><span class="p">(</span><span class="n">stream</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">)]</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">level</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="nb">format</span><span class="p">,</span> <span class="n">handlers</span><span class="o">=</span><span class="n">handlers</span><span class="p">)</span>
    <span class="n">log</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;cisTarget&#39;</span><span class="p">)</span>
    
    <span class="c1"># Load database</span>
    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">ctx_db</span><span class="p">,</span> <span class="nb">str</span><span class="p">):</span>
        <span class="n">ctx_db</span> <span class="o">=</span> <span class="n">cisTargetDatabase</span><span class="p">(</span><span class="n">ctx_db</span><span class="p">,</span>
                             <span class="n">region_sets</span><span class="p">,</span>
                             <span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="p">,</span>
                             <span class="n">fraction_overlap</span> <span class="o">=</span> <span class="n">fraction_overlap</span><span class="p">)</span>
    
    <span class="c1"># Run cistarget analysis in parallel</span>
    <span class="k">if</span> <span class="n">n_cpu</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
        <span class="n">ray</span><span class="o">.</span><span class="n">init</span><span class="p">(</span><span class="n">num_cpus</span><span class="o">=</span><span class="n">n_cpu</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span> <span class="o">=</span> <span class="n">null</span>
        <span class="n">ctx_dict</span> <span class="o">=</span> <span class="n">ray</span><span class="o">.</span><span class="n">get</span><span class="p">([</span><span class="n">ctx_internal_ray</span><span class="o">.</span><span class="n">remote</span><span class="p">(</span><span class="n">ctx_db</span> <span class="o">=</span> <span class="n">ctx_db</span><span class="p">,</span> 
                                            <span class="n">region_set</span> <span class="o">=</span> <span class="n">region_sets</span><span class="p">[</span><span class="n">key</span><span class="p">],</span> 
                                            <span class="n">name</span> <span class="o">=</span> <span class="n">key</span><span class="p">,</span>  
                                            <span class="n">specie</span> <span class="o">=</span> <span class="n">specie</span><span class="p">,</span>
                                            <span class="n">auc_threshold</span> <span class="o">=</span> <span class="n">auc_threshold</span><span class="p">,</span> 
                                            <span class="n">nes_threshold</span> <span class="o">=</span> <span class="n">nes_threshold</span><span class="p">,</span> 
                                            <span class="n">rank_threshold</span> <span class="o">=</span> <span class="n">rank_threshold</span><span class="p">,</span>
                                            <span class="n">path_to_motif_annotations</span> <span class="o">=</span> <span class="n">path_to_motif_annotations</span><span class="p">,</span>
                                            <span class="n">annotation_version</span> <span class="o">=</span> <span class="n">annotation_version</span><span class="p">,</span>
                                            <span class="n">annotation</span> <span class="o">=</span> <span class="n">annotation</span><span class="p">,</span>
                                            <span class="n">motif_similarity_fdr</span> <span class="o">=</span> <span class="n">motif_similarity_fdr</span><span class="p">,</span>
                                            <span class="n">orthologous_identity_threshold</span> <span class="o">=</span> <span class="n">orthologous_identity_threshold</span><span class="p">,</span>
                                            <span class="n">motifs_to_use</span> <span class="o">=</span> <span class="n">motifs_to_use</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">region_sets</span><span class="o">.</span><span class="n">keys</span><span class="p">())])</span>
        <span class="n">ray</span><span class="o">.</span><span class="n">shutdown</span><span class="p">()</span>
        <span class="n">sys</span><span class="o">.</span><span class="n">stderr</span> <span class="o">=</span> <span class="n">sys</span><span class="o">.</span><span class="n">__stderr__</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ctx_dict</span> <span class="o">=</span> <span class="p">[</span><span class="n">ctx_internal</span><span class="p">(</span><span class="n">ctx_db</span> <span class="o">=</span> <span class="n">ctx_db</span><span class="p">,</span> 
                                            <span class="n">region_set</span> <span class="o">=</span> <span class="n">region_sets</span><span class="p">[</span><span class="n">key</span><span class="p">],</span> 
                                            <span class="n">name</span> <span class="o">=</span> <span class="n">key</span><span class="p">,</span>  
                                            <span class="n">specie</span> <span class="o">=</span> <span class="n">specie</span><span class="p">,</span>
                                            <span class="n">auc_threshold</span> <span class="o">=</span> <span class="n">auc_threshold</span><span class="p">,</span> 
                                            <span class="n">nes_threshold</span> <span class="o">=</span> <span class="n">nes_threshold</span><span class="p">,</span> 
                                            <span class="n">rank_threshold</span> <span class="o">=</span> <span class="n">rank_threshold</span><span class="p">,</span>
                                            <span class="n">path_to_motif_annotations</span> <span class="o">=</span> <span class="n">path_to_motif_annotations</span><span class="p">,</span>
                                            <span class="n">annotation_version</span> <span class="o">=</span> <span class="n">annotation_version</span><span class="p">,</span>
                                            <span class="n">annotation</span> <span class="o">=</span> <span class="n">annotation</span><span class="p">,</span>
                                            <span class="n">motif_similarity_fdr</span> <span class="o">=</span> <span class="n">motif_similarity_fdr</span><span class="p">,</span>
                                            <span class="n">orthologous_identity_threshold</span> <span class="o">=</span> <span class="n">orthologous_identity_threshold</span><span class="p">,</span>
                                            <span class="n">motifs_to_use</span> <span class="o">=</span> <span class="n">motifs_to_use</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="nb">list</span><span class="p">(</span><span class="n">region_sets</span><span class="o">.</span><span class="n">keys</span><span class="p">())]</span>
    <span class="n">ctx_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">ctx_result</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">ctx_result</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">region_sets</span><span class="o">.</span><span class="n">keys</span><span class="p">()),</span> <span class="n">ctx_dict</span><span class="p">)}</span>
    <span class="n">log</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s1">&#39;Done!&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ctx_dict</span></div>
        
<span class="nd">@ray</span><span class="o">.</span><span class="n">remote</span>
<span class="k">def</span> <span class="nf">ctx_internal_ray</span><span class="p">(</span><span class="n">ctx_db</span><span class="p">:</span> <span class="n">cisTargetDatabase</span><span class="p">,</span>
            <span class="n">region_set</span><span class="p">:</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">,</span>
            <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">specie</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">auc_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.005</span><span class="p">,</span>
            <span class="n">nes_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
            <span class="n">rank_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
            <span class="n">path_to_motif_annotations</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">annotation_version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;v9&#39;</span><span class="p">,</span>
            <span class="n">annotation</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Direct_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_similarity_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Orthology_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_similarity_and_Orthology_annot&#39;</span><span class="p">],</span>
            <span class="n">motif_similarity_fdr</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
            <span class="n">orthologous_identity_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="n">motifs_to_use</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="kc">None</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
            
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Internal function to run cistarget in parallel with Ray.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ---------</span>
<span class="sd">    ctx_db: :class:`cisTargetDatabase`</span>
<span class="sd">        A cistarget database object.</span>
<span class="sd">    name: str</span>
<span class="sd">        Analysis name</span>
<span class="sd">    specie: str</span>
<span class="sd">        Specie from which genomic coordinates come from</span>
<span class="sd">    auc_threshold: float, optional</span>
<span class="sd">        The fraction of the ranked genome to take into account for the calculation of the</span>
<span class="sd">        Area Under the recovery Curve. Default: 0.005</span>
<span class="sd">    nes_threshold: float, optional</span>
<span class="sd">        The Normalized Enrichment Score (NES) threshold to select enriched features. Default: 3.0</span>
<span class="sd">    rank_threshold: float, optional</span>
<span class="sd">        The total number of ranked genes to take into account when creating a recovery curve.</span>
<span class="sd">        Default: 0.05</span>
<span class="sd">    path_to_motif_annotations: str, optional</span>
<span class="sd">        Path to motif annotations. If not provided, they will be downloaded from </span>
<span class="sd">        https://resources.aertslab.org based on the specie name provided (only possible for mus_musculus,</span>
<span class="sd">        homo_sapiens and drosophila_melanogaster). Default: None</span>
<span class="sd">    annotation_version: str, optional</span>
<span class="sd">        Motif collection version. Default: v9</span>
<span class="sd">    annotation: List, optional</span>
<span class="sd">        Annotation to use for forming cistromes. It can be &#39;Direct_annot&#39; (direct evidence that the motif is </span>
<span class="sd">        linked to that TF), &#39;Motif_similarity_annot&#39; (based on tomtom motif similarity), &#39;Orthology_annot&#39;</span>
<span class="sd">        (based on orthology with a TF that is directly linked to that motif) or &#39;Motif_similarity_and_Orthology_annot&#39;.</span>
<span class="sd">        Default: [&#39;Direct_annot&#39;, &#39;Motif_similarity_annot&#39;, &#39;Orthology_annot&#39;, &#39;Motif_similarity_and_Orthology_annot&#39;]</span>
<span class="sd">    motif_similarity_fdr: float, optional</span>
<span class="sd">        Minimal motif similarity value to consider two motifs similar. Default: 0.001</span>
<span class="sd">    orthologous_identity_threshold: float, optional</span>
<span class="sd">        Minimal orthology value for considering two TFs orthologous. Default: 0.0</span>
<span class="sd">    motifs_to_use: List, optional</span>
<span class="sd">        A subset of motifs to use for the analysis. Default: None (All)</span>
<span class="sd">        </span>
<span class="sd">    Return</span>
<span class="sd">    ---------</span>
<span class="sd">        A :class:`cisTarget` object.</span>
<span class="sd">        </span>
<span class="sd">    References</span>
<span class="sd">    ---------</span>
<span class="sd">    Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.</span>
<span class="sd">    Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">ctx_internal</span><span class="p">(</span><span class="n">ctx_db</span> <span class="o">=</span> <span class="n">ctx_db</span><span class="p">,</span>
                        <span class="n">region_set</span> <span class="o">=</span> <span class="n">region_set</span><span class="p">,</span>
                        <span class="n">name</span> <span class="o">=</span> <span class="n">name</span><span class="p">,</span>
                        <span class="n">specie</span> <span class="o">=</span> <span class="n">specie</span><span class="p">,</span>
                        <span class="n">auc_threshold</span> <span class="o">=</span> <span class="n">auc_threshold</span><span class="p">,</span>
                        <span class="n">nes_threshold</span> <span class="o">=</span> <span class="n">nes_threshold</span><span class="p">,</span>
                        <span class="n">rank_threshold</span> <span class="o">=</span> <span class="n">rank_threshold</span><span class="p">,</span>
                        <span class="n">path_to_motif_annotations</span> <span class="o">=</span> <span class="n">path_to_motif_annotations</span><span class="p">,</span>
                        <span class="n">annotation_version</span> <span class="o">=</span> <span class="n">annotation_version</span><span class="p">,</span>
                        <span class="n">annotation</span> <span class="o">=</span> <span class="n">annotation</span><span class="p">,</span>
                        <span class="n">motif_similarity_fdr</span> <span class="o">=</span> <span class="n">motif_similarity_fdr</span><span class="p">,</span>
                        <span class="n">orthologous_identity_threshold</span> <span class="o">=</span> <span class="n">orthologous_identity_threshold</span><span class="p">,</span>
                        <span class="n">motifs_to_use</span> <span class="o">=</span> <span class="n">motifs_to_use</span><span class="p">)</span>


<div class="viewcode-block" id="ctx_internal"><a class="viewcode-back" href="../../api.html#pycistarget.motif_enrichment_cistarget.ctx_internal">[docs]</a><span class="k">def</span> <span class="nf">ctx_internal</span><span class="p">(</span><span class="n">ctx_db</span><span class="p">:</span> <span class="n">cisTargetDatabase</span><span class="p">,</span>
            <span class="n">region_set</span><span class="p">:</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">,</span>
            <span class="n">name</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">specie</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">auc_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.005</span><span class="p">,</span>
            <span class="n">nes_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
            <span class="n">rank_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
            <span class="n">path_to_motif_annotations</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">annotation_version</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s1">&#39;v9&#39;</span><span class="p">,</span>
            <span class="n">annotation</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Direct_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_similarity_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Orthology_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Motif_similarity_and_Orthology_annot&#39;</span><span class="p">],</span>
            <span class="n">motif_similarity_fdr</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.001</span><span class="p">,</span>
            <span class="n">orthologous_identity_threshold</span><span class="p">:</span> <span class="nb">float</span> <span class="o">=</span> <span class="mf">0.0</span><span class="p">,</span>
            <span class="n">motifs_to_use</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="kc">None</span> <span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Internal function to run cistarget.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ---------</span>
<span class="sd">    ctx_db: :class:`cisTargetDatabase`</span>
<span class="sd">        A cistarget database object.</span>
<span class="sd">    name: str</span>
<span class="sd">        Analysis name</span>
<span class="sd">    specie: str</span>
<span class="sd">        Specie from which genomic coordinates come from</span>
<span class="sd">    auc_threshold: float, optional</span>
<span class="sd">        The fraction of the ranked genome to take into account for the calculation of the</span>
<span class="sd">        Area Under the recovery Curve. Default: 0.005</span>
<span class="sd">    nes_threshold: float, optional</span>
<span class="sd">        The Normalized Enrichment Score (NES) threshold to select enriched features. Default: 3.0</span>
<span class="sd">    rank_threshold: float, optional</span>
<span class="sd">        The total number of ranked genes to take into account when creating a recovery curve.</span>
<span class="sd">        Default: 0.05</span>
<span class="sd">    path_to_motif_annotations: str, optional</span>
<span class="sd">        Path to motif annotations. If not provided, they will be downloaded from </span>
<span class="sd">        https://resources.aertslab.org based on the specie name provided (only possible for mus_musculus,</span>
<span class="sd">        homo_sapiens and drosophila_melanogaster). Default: None</span>
<span class="sd">    annotation_version: str, optional</span>
<span class="sd">        Motif collection version. Default: v9</span>
<span class="sd">    annotation: List, optional</span>
<span class="sd">        Annotation to use for forming cistromes. It can be &#39;Direct_annot&#39; (direct evidence that the motif is </span>
<span class="sd">        linked to that TF), &#39;Motif_similarity_annot&#39; (based on tomtom motif similarity), &#39;Orthology_annot&#39;</span>
<span class="sd">        (based on orthology with a TF that is directly linked to that motif) or &#39;Motif_similarity_and_Orthology_annot&#39;.</span>
<span class="sd">        Default: [&#39;Direct_annot&#39;, &#39;Motif_similarity_annot&#39;, &#39;Orthology_annot&#39;, &#39;Motif_similarity_and_Orthology_annot&#39;]</span>
<span class="sd">    motif_similarity_fdr: float, optional</span>
<span class="sd">        Minimal motif similarity value to consider two motifs similar. Default: 0.001</span>
<span class="sd">    orthologous_identity_threshold: float, optional</span>
<span class="sd">        Minimal orthology value for considering two TFs orthologous. Default: 0.0</span>
<span class="sd">    motifs_to_use: List, optional</span>
<span class="sd">        A subset of motifs to use for the analysis. Default: None (All)</span>
<span class="sd">        </span>
<span class="sd">    Return</span>
<span class="sd">    ---------</span>
<span class="sd">        A :class:`cisTarget` object.</span>
<span class="sd">        </span>
<span class="sd">    References</span>
<span class="sd">    ---------</span>
<span class="sd">    Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.</span>
<span class="sd">    Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ctx_result</span> <span class="o">=</span> <span class="n">cisTarget</span><span class="p">(</span><span class="n">region_set</span><span class="p">,</span> 
                           <span class="n">name</span><span class="p">,</span> 
                           <span class="n">specie</span><span class="p">,</span>
                           <span class="n">auc_threshold</span><span class="p">,</span>
                           <span class="n">nes_threshold</span><span class="p">,</span>
                           <span class="n">rank_threshold</span><span class="p">,</span>
                           <span class="n">path_to_motif_annotations</span><span class="p">,</span>
                           <span class="n">annotation_version</span><span class="p">,</span>
                           <span class="n">annotation</span><span class="p">,</span>
                           <span class="n">motif_similarity_fdr</span><span class="p">,</span>
                           <span class="n">orthologous_identity_threshold</span><span class="p">,</span>
                           <span class="n">motifs_to_use</span><span class="p">)</span>
    <span class="n">ctx_result</span><span class="o">.</span><span class="n">run_ctx</span><span class="p">(</span><span class="n">ctx_db</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">ctx_result</span></div>
    
<span class="c1">## Show results </span>
<div class="viewcode-block" id="cistarget_results"><a class="viewcode-back" href="../../api.html#pycistarget.motif_enrichment_cistarget.cistarget_results">[docs]</a><span class="k">def</span> <span class="nf">cistarget_results</span><span class="p">(</span><span class="n">cistarget_dict</span><span class="p">,</span>
                    <span class="n">name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    A function to show cistarget results in jupyter notebooks.</span>
<span class="sd">    </span>
<span class="sd">    Parameters</span>
<span class="sd">    ---------</span>
<span class="sd">    cistarget_dict: Dict</span>
<span class="sd">        A dictionary with one :class:`cisTarget` object per slot.</span>
<span class="sd">    name: str</span>
<span class="sd">        Dictionary key of the analysis result to show. Default: None (All)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">motif_enrichment_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">cistarget_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">motif_enrichment</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">cistarget_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
    <span class="k">if</span> <span class="n">name</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">motif_enrichment_table</span><span class="o">=</span><span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">motif_enrichment_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">motif_enrichment_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">motif_enrichment_table</span><span class="o">=</span><span class="n">motif_enrichment_dict</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">HTML</span><span class="p">(</span><span class="n">motif_enrichment_table</span><span class="o">.</span><span class="n">to_html</span><span class="p">(</span><span class="n">escape</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">col_space</span><span class="o">=</span><span class="mi">80</span><span class="p">))</span></div>


</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Carmen Bravo Gonzalez Blas, Seppe de Winter.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>