<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API &mdash; pycistarget 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="PycisTarget on human cerebellum DARs and topics" href="pycistarget_scenic%2B_human_brain.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> pycistarget
          </a>
              <div class="version">
                1.0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tools.html">pycisTarget methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="motif_table.html">SCENIC+ motif collection</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#module-pycistarget.motif_enrichment_cistarget">cisTarget</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pycistarget.motif_enrichment_dem">DEM</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pycistarget.motif_enrichment_homer">Homer</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pycistarget.cluster_buster">Cluster-Buster</a></li>
<li class="toctree-l2"><a class="reference internal" href="#module-pycistarget.utils">Utils</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pycistarget</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>API</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/api.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="api">
<span id="id1"></span><h1>API<a class="headerlink" href="#api" title="Permalink to this heading"></a></h1>
<div class="section" id="module-pycistarget.motif_enrichment_cistarget">
<span id="cistarget"></span><h2>cisTarget<a class="headerlink" href="#module-pycistarget.motif_enrichment_cistarget" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_cistarget.</span></span><span class="sig-name descname"><span class="pre">cisTarget</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PyRanges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specie</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auc_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.005</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nes_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motif_annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'v9'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_similarity_fdr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orthologous_identity_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motifs_to_use</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_cistarget.html#cisTarget"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget" title="Permalink to this definition"></a></dt>
<dd><p>cisTarget class.
<a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTarget" title="pycistarget.motif_enrichment_cistarget.cisTarget"><code class="xref py py-class docutils literal notranslate"><span class="pre">cisTarget</span></code></a> contains method for motif enrichment analysis on sets of regions.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.regions_to_db">
<span class="sig-name descname"><span class="pre">regions_to_db</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.regions_to_db" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe containing the mapping between query regions and regions in the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.region_set">
<span class="sig-name descname"><span class="pre">region_set</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.region_set" title="Permalink to this definition"></a></dt>
<dd><p>A PyRanges containing region coordinates for the regions to be analyzed.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pr.PyRanges</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.name" title="Permalink to this definition"></a></dt>
<dd><p>Analysis name</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.specie">
<span class="sig-name descname"><span class="pre">specie</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.specie" title="Permalink to this definition"></a></dt>
<dd><p>Specie from which genomic coordinates come from</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.auc_threshold">
<span class="sig-name descname"><span class="pre">auc_threshold</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.auc_threshold" title="Permalink to this definition"></a></dt>
<dd><p>The fraction of the ranked genome to take into account for the calculation of the
Area Under the recovery Curve. Default: 0.005</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.nes_threshold">
<span class="sig-name descname"><span class="pre">nes_threshold</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.nes_threshold" title="Permalink to this definition"></a></dt>
<dd><p>The Normalized Enrichment Score (NES) threshold to select enriched features. Default: 3.0</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.rank_threshold">
<span class="sig-name descname"><span class="pre">rank_threshold</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.rank_threshold" title="Permalink to this definition"></a></dt>
<dd><p>The total number of ranked genes to take into account when creating a recovery curve.
Default: 0.05</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.path_to_motif_annotations">
<span class="sig-name descname"><span class="pre">path_to_motif_annotations</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.path_to_motif_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Path to motif annotations. If not provided, they will be downloaded from
<a class="reference external" href="https://resources.aertslab.org">https://resources.aertslab.org</a> based on the specie name provided (only possible for mus_musculus,
homo_sapiens and drosophila_melanogaster). Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.annotation_version">
<span class="sig-name descname"><span class="pre">annotation_version</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.annotation_version" title="Permalink to this definition"></a></dt>
<dd><p>Motif collection version. Default: v9</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.annotation">
<span class="sig-name descname"><span class="pre">annotation</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.annotation" title="Permalink to this definition"></a></dt>
<dd><p>Annotation to use for forming cistromes. It can be ‘Direct_annot’ (direct evidence that the motif is
linked to that TF), ‘Motif_similarity_annot’ (based on tomtom motif similarity), ‘Orthology_annot’
(based on orthology with a TF that is directly linked to that motif) or ‘Motif_similarity_and_Orthology_annot’.
Default: [‘Direct_annot’, ‘Motif_similarity_annot’, ‘Orthology_annot’, ‘Motif_similarity_and_Orthology_annot’]</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>List, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.motif_similarity_fdr">
<span class="sig-name descname"><span class="pre">motif_similarity_fdr</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.motif_similarity_fdr" title="Permalink to this definition"></a></dt>
<dd><p>Minimal motif similarity value to consider two motifs similar. Default: 0.001</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.orthologous_identity_threshold">
<span class="sig-name descname"><span class="pre">orthologous_identity_threshold</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.orthologous_identity_threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimal orthology value for considering two TFs orthologous. Default: 0.0</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.motifs_to_use">
<span class="sig-name descname"><span class="pre">motifs_to_use</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.motifs_to_use" title="Permalink to this definition"></a></dt>
<dd><p>A subset of motifs to use for the analysis. Default: None (All)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>List, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.motif_enrichment">
<span class="sig-name descname"><span class="pre">motif_enrichment</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.motif_enrichment" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe containing motif enrichment results</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.motif_hits">
<span class="sig-name descname"><span class="pre">motif_hits</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.motif_hits" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary containing regions that are considered enriched for each motif.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.cistromes">
<span class="sig-name descname"><span class="pre">cistromes</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.cistromes" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary containing TF cistromes. Cistromes with no extension contain regions linked to directly
annotated motifs, while ‘_extended’ cistromes can contain regions linked to motifs annotated by
similarity or orthology.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">References</p>
<p>Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.
Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</p>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTarget.add_motif_annotation_cistarget" title="pycistarget.motif_enrichment_cistarget.cisTarget.add_motif_annotation_cistarget"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_motif_annotation_cistarget</span></code></a>([add_logo])</p></td>
<td><p>Add motif annotation</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTarget.run_ctx" title="pycistarget.motif_enrichment_cistarget.cisTarget.run_ctx"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run_ctx</span></code></a>(ctx_db)</p></td>
<td><p>Run cisTarget</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.add_motif_annotation_cistarget">
<span class="sig-name descname"><span class="pre">add_motif_annotation_cistarget</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">add_logo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_cistarget.html#cisTarget.add_motif_annotation_cistarget"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.add_motif_annotation_cistarget" title="Permalink to this definition"></a></dt>
<dd><p>Add motif annotation</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>add_logo</strong> (<em>boolean</em><em>, </em><em>optional</em>) – Whether to add the motif logo to the motif enrichment dataframe</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.
Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTarget.run_ctx">
<span class="sig-name descname"><span class="pre">run_ctx</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ctx_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase" title="pycistarget.motif_enrichment_cistarget.cisTargetDatabase"><span class="pre">cisTargetDatabase</span></a></span></em><span class="sig-paren">)</span> <span class="sig-return"><span class="sig-return-icon">&#x2192;</span> <span class="sig-return-typehint"><span class="pre">DataFrame</span></span></span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_cistarget.html#cisTarget.run_ctx"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTarget.run_ctx" title="Permalink to this definition"></a></dt>
<dd><p>Run cisTarget</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ctx_db</strong> (<a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase" title="pycistarget.motif_enrichment_cistarget.cisTargetDatabase"><code class="xref py py-class docutils literal notranslate"><span class="pre">cisTargetDatabase</span></code></a>) – A cistarget database object.</p>
</dd>
</dl>
<p class="rubric">References</p>
<p>Van de Sande B., Flerin C., et al. A scalable SCENIC workflow for single-cell gene regulatory network analysis.
Nat Protoc. June 2020:1-30. doi:10.1038/s41596-020-0336-2</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTargetDatabase">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_cistarget.</span></span><span class="sig-name descname"><span class="pre">cisTargetDatabase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_cistarget.html#cisTargetDatabase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase" title="Permalink to this definition"></a></dt>
<dd><p>cisTarget Database class.
<a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase" title="pycistarget.motif_enrichment_cistarget.cisTargetDatabase"><code class="xref py py-class docutils literal notranslate"><span class="pre">cisTargetDatabase</span></code></a> contains a dataframe with motifs as rows, regions as columns and rank as
values. In addition, is contains a slot to map query regions to regions in the database. For more
information on how to generate databases, please visit: <a class="reference external" href="https://github.com/aertslab/create_cisTarget_databases">https://github.com/aertslab/create_cisTarget_databases</a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTargetDatabase.regions_to_db">
<span class="sig-name descname"><span class="pre">regions_to_db</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase.regions_to_db" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe containing the mapping between query regions and regions in the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTargetDatabase.db_rankings">
<span class="sig-name descname"><span class="pre">db_rankings</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase.db_rankings" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe with motifs as rows, regions as columns and rank as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTargetDatabase.total_regions">
<span class="sig-name descname"><span class="pre">total_regions</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase.total_regions" title="Permalink to this definition"></a></dt>
<dd><p>Total number of regions in the database</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase.load_db" title="pycistarget.motif_enrichment_cistarget.cisTargetDatabase.load_db"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_db</span></code></a>(fname[, region_sets, name, ...])</p></td>
<td><p>Load cisTargetDatabase</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cisTargetDatabase.load_db">
<span class="sig-name descname"><span class="pre">load_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_cistarget.html#cisTargetDatabase.load_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase.load_db" title="Permalink to this definition"></a></dt>
<dd><p>Load cisTargetDatabase</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> (<em>str</em>) – Path to feather file containing the cisTarget database (regions_vs_motifs)</p></li>
<li><p><strong>region_sets</strong> (<em>Dict</em><em> or </em><em>pr.PyRanges</em><em>, </em><em>optional</em>) – Dictionary or pr.PyRanges that are going to be analyzed with cistarget. Default: None.</p></li>
<li><p><strong>name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name for the cistarget database. Default: None</p></li>
<li><p><strong>fraction_overlap</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal overlap between query and regions in the database for the mapping.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.cistarget_results">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_cistarget.</span></span><span class="sig-name descname"><span class="pre">cistarget_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cistarget_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_cistarget.html#cistarget_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.cistarget_results" title="Permalink to this definition"></a></dt>
<dd><p>A function to show cistarget results in jupyter notebooks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cistarget_dict</strong> (<em>Dict</em>) – A dictionary with one <a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTarget" title="pycistarget.motif_enrichment_cistarget.cisTarget"><code class="xref py py-class docutils literal notranslate"><span class="pre">cisTarget</span></code></a> object per slot.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Dictionary key of the analysis result to show. Default: None (All)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.ctx_internal">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_cistarget.</span></span><span class="sig-name descname"><span class="pre">ctx_internal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ctx_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase" title="pycistarget.motif_enrichment_cistarget.cisTargetDatabase"><span class="pre">cisTargetDatabase</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PyRanges</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specie</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auc_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.005</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nes_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motif_annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'v9'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_similarity_fdr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orthologous_identity_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motifs_to_use</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_cistarget.html#ctx_internal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.ctx_internal" title="Permalink to this definition"></a></dt>
<dd><p>Internal function to run cistarget.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ctx_db</strong> (<a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase" title="pycistarget.motif_enrichment_cistarget.cisTargetDatabase"><code class="xref py py-class docutils literal notranslate"><span class="pre">cisTargetDatabase</span></code></a>) – A cistarget database object.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Analysis name</p></li>
<li><p><strong>specie</strong> (<em>str</em>) – Specie from which genomic coordinates come from</p></li>
<li><p><strong>auc_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The fraction of the ranked genome to take into account for the calculation of the
Area Under the recovery Curve. Default: 0.005</p></li>
<li><p><strong>nes_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The Normalized Enrichment Score (NES) threshold to select enriched features. Default: 3.0</p></li>
<li><p><strong>rank_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The total number of ranked genes to take into account when creating a recovery curve.
Default: 0.05</p></li>
<li><p><strong>path_to_motif_annotations</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to motif annotations. If not provided, they will be downloaded from
<a class="reference external" href="https://resources.aertslab.org">https://resources.aertslab.org</a> based on the specie name provided (only possible for mus_musculus,
homo_sapiens and drosophila_melanogaster). Default: None</p></li>
<li><p><strong>annotation_version</strong> (<em>str</em><em>, </em><em>optional</em>) – Motif collection version. Default: v9</p></li>
<li><p><strong>annotation</strong> (<em>List</em><em>, </em><em>optional</em>) – Annotation to use for forming cistromes. It can be ‘Direct_annot’ (direct evidence that the motif is
linked to that TF), ‘Motif_similarity_annot’ (based on tomtom motif similarity), ‘Orthology_annot’
(based on orthology with a TF that is directly linked to that motif) or ‘Motif_similarity_and_Orthology_annot’.
Default: [‘Direct_annot’, ‘Motif_similarity_annot’, ‘Orthology_annot’, ‘Motif_similarity_and_Orthology_annot’]</p></li>
<li><p><strong>motif_similarity_fdr</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal motif similarity value to consider two motifs similar. Default: 0.001</p></li>
<li><p><strong>orthologous_identity_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal orthology value for considering two TFs orthologous. Default: 0.0</p></li>
<li><p><strong>motifs_to_use</strong> (<em>List</em><em>, </em><em>optional</em>) – A subset of motifs to use for the analysis. Default: None (All)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_cistarget.run_cistarget">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_cistarget.</span></span><span class="sig-name descname"><span class="pre">run_cistarget</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ctx_db</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase" title="pycistarget.motif_enrichment_cistarget.cisTargetDatabase"><span class="pre">cisTargetDatabase</span></a></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specie</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'cisTarget'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">auc_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.005</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nes_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">3.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rank_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motif_annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'v9'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_similarity_fdr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orthologous_identity_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cpu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motifs_to_use</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_cistarget.html#run_cistarget"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_cistarget.run_cistarget" title="Permalink to this definition"></a></dt>
<dd><p>Run cisTarget.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ctx_db</strong> (<a class="reference internal" href="#pycistarget.motif_enrichment_cistarget.cisTargetDatabase" title="pycistarget.motif_enrichment_cistarget.cisTargetDatabase"><code class="xref py py-class docutils literal notranslate"><span class="pre">cisTargetDatabase</span></code></a>) – A cistarget database object.</p></li>
<li><p><strong>region_sets</strong> (<em>Dict</em>) – A dictionary of PyRanges containing region coordinates for the region sets to be analyzed.</p></li>
<li><p><strong>specie</strong> (<em>str</em>) – Specie from which genomic coordinates come from</p></li>
<li><p><strong>name</strong> (<em>str</em><em>, </em><em>optional</em>) – Analysis name. Default: cisTarget</p></li>
<li><p><strong>fraction_overlap</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal overlap between query and regions in the database for the mapping. Default: 0.4</p></li>
<li><p><strong>auc_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The fraction of the ranked genome to take into account for the calculation of the
Area Under the recovery Curve. Default: 0.005</p></li>
<li><p><strong>nes_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The Normalized Enrichment Score (NES) threshold to select enriched features. Default: 3.0</p></li>
<li><p><strong>rank_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – The total number of ranked genes to take into account when creating a recovery curve.
Default: 0.05</p></li>
<li><p><strong>path_to_motif_annotations</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to motif annotations. If not provided, they will be downloaded from
<a class="reference external" href="https://resources.aertslab.org">https://resources.aertslab.org</a> based on the specie name provided (only possible for mus_musculus,
homo_sapiens and drosophila_melanogaster). Default: None</p></li>
<li><p><strong>annotation_version</strong> (<em>str</em><em>, </em><em>optional</em>) – Motif collection version. Default: v9</p></li>
<li><p><strong>annotation</strong> (<em>List</em><em>, </em><em>optional</em>) – Annotation to use for forming cistromes. It can be ‘Direct_annot’ (direct evidence that the motif is
linked to that TF), ‘Motif_similarity_annot’ (based on tomtom motif similarity), ‘Orthology_annot’
(based on orthology with a TF that is directly linked to that motif) or ‘Motif_similarity_and_Orthology_annot’.
Default: [‘Direct_annot’, ‘Motif_similarity_annot’, ‘Orthology_annot’, ‘Motif_similarity_and_Orthology_annot’]</p></li>
<li><p><strong>motif_similarity_fdr</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal motif similarity value to consider two motifs similar. Default: 0.001</p></li>
<li><p><strong>orthologous_identity_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal orthology value for considering two TFs orthologous. Default: 0.0</p></li>
<li><p><strong>n_cpu</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of cores to use. If 1, ray will not be used. Default: 1</p></li>
<li><p><strong>motifs_to_use</strong> (<em>List</em><em>, </em><em>optional</em>) – A subset of motifs to use for the analysis. Default: None (All)</p></li>
<li><p><strong>**kwargs</strong> – Extra parameters to pass to <cite>ray.init()</cite></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-pycistarget.motif_enrichment_dem">
<span id="dem"></span><h2>DEM<a class="headerlink" href="#module-pycistarget.motif_enrichment_dem" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_dem.</span></span><span class="sig-name descname"><span class="pre">DEM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dem_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specie</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset_motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contrasts</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'Other'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'DEM'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_bg_regions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjpval_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log2fc_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_fg_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_hit_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cpu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cluster_buster_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_genome_fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">promoter_space</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motif_annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'v9'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_similarity_fdr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orthologous_identity_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tmp_dir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#DEM"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM" title="Permalink to this definition"></a></dt>
<dd><p>DEM class.
<a class="reference internal" href="#pycistarget.motif_enrichment_dem.DEM" title="pycistarget.motif_enrichment_dem.DEM"><code class="xref py py-class docutils literal notranslate"><span class="pre">DEM</span></code></a> contains DEM method for motif enrichment analysis on sets of regions.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.regions_to_db">
<span class="sig-name descname"><span class="pre">regions_to_db</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.regions_to_db" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe containing the mapping between query regions and regions in the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.region_sets">
<span class="sig-name descname"><span class="pre">region_sets</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.region_sets" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary of PyRanges containing region coordinates for the regions to be analyzed.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.specie">
<span class="sig-name descname"><span class="pre">specie</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.specie" title="Permalink to this definition"></a></dt>
<dd><p>Specie from which genomic coordinates come from.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.subset_motifs">
<span class="sig-name descname"><span class="pre">subset_motifs</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.subset_motifs" title="Permalink to this definition"></a></dt>
<dd><p>List of motifs to disregard in the analysis. Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>List, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.contrasts">
<span class="sig-name descname"><span class="pre">contrasts</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.contrasts" title="Permalink to this definition"></a></dt>
<dd><p>List of contrasts to perform. Default: None (Each group versus all the rest)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>List, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.name" title="Permalink to this definition"></a></dt>
<dd><p>Analysis name</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.max_bg_regions">
<span class="sig-name descname"><span class="pre">max_bg_regions</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.max_bg_regions" title="Permalink to this definition"></a></dt>
<dd><p>Maximum number of regions to use as background. Default: None (All)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.adjpval_thr">
<span class="sig-name descname"><span class="pre">adjpval_thr</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.adjpval_thr" title="Permalink to this definition"></a></dt>
<dd><p>Adjusted p-value threshold to consider a motif enriched. Default: 0.05</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.log2fc_thr">
<span class="sig-name descname"><span class="pre">log2fc_thr</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.log2fc_thr" title="Permalink to this definition"></a></dt>
<dd><p>Log2 Fold-change threshold to consider a motif enriched. Default: 1</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.mean_fg_thr">
<span class="sig-name descname"><span class="pre">mean_fg_thr</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.mean_fg_thr" title="Permalink to this definition"></a></dt>
<dd><p>Minimul mean signal in the foreground to consider a motif enriched. Default: 0</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.motif_hit_thr">
<span class="sig-name descname"><span class="pre">motif_hit_thr</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.motif_hit_thr" title="Permalink to this definition"></a></dt>
<dd><p>Minimal CRM score to consider a region enriched for a motif. Default: None (It will be automatically
calculated based on precision-recall).</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.n_cpu">
<span class="sig-name descname"><span class="pre">n_cpu</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.n_cpu" title="Permalink to this definition"></a></dt>
<dd><p>Number of cores to use. Default: 1</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.fraction_overlap">
<span class="sig-name descname"><span class="pre">fraction_overlap</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.fraction_overlap" title="Permalink to this definition"></a></dt>
<dd><p>Minimal overlap between query and regions in the database for the mapping.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.cluster_buster_path">
<span class="sig-name descname"><span class="pre">cluster_buster_path</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.cluster_buster_path" title="Permalink to this definition"></a></dt>
<dd><p>Path to cluster buster bin. Only required if using a shuffled background. Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.path_to_genome_fasta">
<span class="sig-name descname"><span class="pre">path_to_genome_fasta</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.path_to_genome_fasta" title="Permalink to this definition"></a></dt>
<dd><p>Path to genome fasta file. Only required if using a shuffled background. Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.path_to_motifs">
<span class="sig-name descname"><span class="pre">path_to_motifs</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.path_to_motifs" title="Permalink to this definition"></a></dt>
<dd><p>Path to motif collection folder (in .cb format). Only required if using a shuffled background.
Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.genome_annotation">
<span class="sig-name descname"><span class="pre">genome_annotation</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.genome_annotation" title="Permalink to this definition"></a></dt>
<dd><p>Pyranges containing genome annotation (e.g. biomart). Only required if using promoter balance.
Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pr.PyRanges, optional.</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.promoter_space">
<span class="sig-name descname"><span class="pre">promoter_space</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.promoter_space" title="Permalink to this definition"></a></dt>
<dd><p>Space around TSS to consider a region promoter. Only used if using promoter balance.
Default: 1000</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.path_to_motif_annotations">
<span class="sig-name descname"><span class="pre">path_to_motif_annotations</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.path_to_motif_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Path to motif annotations. If not provided, they will be downloaded from
<a class="reference external" href="https://resources.aertslab.org">https://resources.aertslab.org</a> based on the specie name provided (only possible for mus_musculus,
homo_sapiens and drosophila_melanogaster). Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.motif_similarity_fdr">
<span class="sig-name descname"><span class="pre">motif_similarity_fdr</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.motif_similarity_fdr" title="Permalink to this definition"></a></dt>
<dd><p>Minimal motif similarity value to consider two motifs similar. Default: 0.001</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.orthologous_identity_threshold">
<span class="sig-name descname"><span class="pre">orthologous_identity_threshold</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.orthologous_identity_threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimal orthology value for considering two TFs orthologous. Default: 0.0</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.motifs_to_use">
<span class="sig-name descname"><span class="pre">motifs_to_use</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.motifs_to_use" title="Permalink to this definition"></a></dt>
<dd><p>A subset of motifs to use for the analysis. Default: None (All)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>List, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.tmp_dir">
<span class="sig-name descname"><span class="pre">tmp_dir</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.tmp_dir" title="Permalink to this definition"></a></dt>
<dd><p>Temp directory to use if running cluster_buster. Default: None (        mp)</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.motif_enrichment">
<span class="sig-name descname"><span class="pre">motif_enrichment</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.motif_enrichment" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe containing motif enrichment results</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.motif_hits">
<span class="sig-name descname"><span class="pre">motif_hits</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.motif_hits" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary containing regions that are considered enriched for each motif.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.cistromes">
<span class="sig-name descname"><span class="pre">cistromes</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.cistromes" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary containing TF cistromes. Cistromes with no extension contain regions linked to directly
annotated motifs, while ‘_extended’ cistromes can contain regions linked to motifs annotated by
similarity or orthology.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_dem.DEM.add_motif_annotation_dem" title="pycistarget.motif_enrichment_dem.DEM.add_motif_annotation_dem"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_motif_annotation_dem</span></code></a>([add_logo])</p></td>
<td><p>Add motif annotation</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_dem.DEM.run" title="pycistarget.motif_enrichment_dem.DEM.run"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run</span></code></a>(dem_db_scores, **kwargs)</p></td>
<td><p>Run DEM</p></td>
</tr>
</tbody>
</table>
<table class="docutils align-default">
<colgroup>
<col style="width: 60%" />
<col style="width: 40%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><strong>DEM_results</strong></p></td>
<td></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.add_motif_annotation_dem">
<span class="sig-name descname"><span class="pre">add_motif_annotation_dem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">add_logo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#DEM.add_motif_annotation_dem"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.add_motif_annotation_dem" title="Permalink to this definition"></a></dt>
<dd><p>Add motif annotation</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>add_logo</strong> (<em>boolean</em><em>, </em><em>optional</em>) – Whether to add the motif logo to the motif enrichment dataframe</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dem_db_scores</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#DEM.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM.run" title="Permalink to this definition"></a></dt>
<dd><p>Run DEM</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dem_db_scores</strong> (<em>pd.DataFrame</em>) – A dataframe containing maximum CRM score for each motif in each regions.</p></li>
<li><p><strong>**kwargs</strong> – Additional parameters to pass to <cite>ray.init()</cite></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEMDatabase">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_dem.</span></span><span class="sig-name descname"><span class="pre">DEMDatabase</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#DEMDatabase"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEMDatabase" title="Permalink to this definition"></a></dt>
<dd><p>DEM Database class.
<a class="reference internal" href="#pycistarget.motif_enrichment_dem.DEMDatabase" title="pycistarget.motif_enrichment_dem.DEMDatabase"><code class="xref py py-class docutils literal notranslate"><span class="pre">DEMDatabase</span></code></a> contains a dataframe with motifs as rows, regions as columns and CRM scores as
values. In addition, is contains a slot to map query regions to regions in the database. For more
information on how to generate databases, please visit: <a class="reference external" href="https://github.com/aertslab/create_cisTarget_databases">https://github.com/aertslab/create_cisTarget_databases</a></p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEMDatabase.regions_to_db">
<span class="sig-name descname"><span class="pre">regions_to_db</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEMDatabase.regions_to_db" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe containing the mapping between query regions and regions in the database.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEMDatabase.db_scores">
<span class="sig-name descname"><span class="pre">db_scores</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEMDatabase.db_scores" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe with motifs as rows, regions as columns and CRM scores as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEMDatabase.total_regions">
<span class="sig-name descname"><span class="pre">total_regions</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEMDatabase.total_regions" title="Permalink to this definition"></a></dt>
<dd><p>Total number of regions in the database</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_dem.DEMDatabase.load_db" title="pycistarget.motif_enrichment_dem.DEMDatabase.load_db"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_db</span></code></a>(fname[, region_sets, name, ...])</p></td>
<td><p>Load DEMDatabase</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEMDatabase.load_db">
<span class="sig-name descname"><span class="pre">load_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#DEMDatabase.load_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEMDatabase.load_db" title="Permalink to this definition"></a></dt>
<dd><p>Load DEMDatabase</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> (<em>str</em>) – Path to feather file containing the DEM database (regions_vs_motifs)</p></li>
<li><p><strong>region_sets</strong> (<em>Dict</em><em> or </em><em>pr.PyRanges</em><em>, </em><em>optional</em>) – Dictionary or pr.PyRanges that are going to be analyzed with DEM. Default: None.</p></li>
<li><p><strong>name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name for the DEM database. Default: None</p></li>
<li><p><strong>fraction_overlap</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal overlap between query and regions in the database for the mapping.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.DEM_internal">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_dem.</span></span><span class="sig-name descname"><span class="pre">DEM_internal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dem_db_scores</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_group</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">contrast_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">adjpval_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.05</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log2fc_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mean_fg_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_hit_thr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">float</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#DEM_internal"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.DEM_internal" title="Permalink to this definition"></a></dt>
<dd><p>Internal operations for DEM.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.create_groups">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_dem.</span></span><span class="sig-name descname"><span class="pre">create_groups</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">contrast</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_bg_regions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_genome_fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_regions_fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cbust_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">promoter_space</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cpu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#create_groups"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.create_groups" title="Permalink to this definition"></a></dt>
<dd><p>”
Format contrast groups</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.get_motif_hits">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_dem.</span></span><span class="sig-name descname"><span class="pre">get_motif_hits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">labels</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optimal_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#get_motif_hits"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.get_motif_hits" title="Permalink to this definition"></a></dt>
<dd><p>Determine optimal score threshold based on precision-recall.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.p_adjust_bh">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_dem.</span></span><span class="sig-name descname"><span class="pre">p_adjust_bh</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#p_adjust_bh"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.p_adjust_bh" title="Permalink to this definition"></a></dt>
<dd><p>Benjamini-Hochberg p-value correction for multiple hypothesis testing.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_dem.shuffle_sequence">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_dem.</span></span><span class="sig-name descname"><span class="pre">shuffle_sequence</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sequence</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_dem.html#shuffle_sequence"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_dem.shuffle_sequence" title="Permalink to this definition"></a></dt>
<dd><p>Shuffle given sequence</p>
</dd></dl>

</div>
<div class="section" id="module-pycistarget.motif_enrichment_homer">
<span id="homer"></span><h2>Homer<a class="headerlink" href="#module-pycistarget.motif_enrichment_homer" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_homer.</span></span><span class="sig-name descname"><span class="pre">Homer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">homer_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bed_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'given'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">denovo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'8,10,12'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meme_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meme_collection_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motif_annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'v9'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cistrome_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_similarity_fdr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orthologous_identity_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#Homer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer" title="Permalink to this definition"></a></dt>
<dd><p>Homer class.
<a class="reference internal" href="#pycistarget.motif_enrichment_homer.Homer" title="pycistarget.motif_enrichment_homer.Homer"><code class="xref py py-class docutils literal notranslate"><span class="pre">Homer</span></code></a> contains Homer for motif enrichment analysis on sets of regions.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.homer_path">
<span class="sig-name descname"><span class="pre">homer_path</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.homer_path" title="Permalink to this definition"></a></dt>
<dd><p>Path to Homer bin folder.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.bed_path">
<span class="sig-name descname"><span class="pre">bed_path</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.bed_path" title="Permalink to this definition"></a></dt>
<dd><p>Path to bed file containing region set to be analyzed with Homer.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.name">
<span class="sig-name descname"><span class="pre">name</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.name" title="Permalink to this definition"></a></dt>
<dd><p>Analysis name.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.outdir">
<span class="sig-name descname"><span class="pre">outdir</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.outdir" title="Permalink to this definition"></a></dt>
<dd><p>Path to folder to output Homer results.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.genome">
<span class="sig-name descname"><span class="pre">genome</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.genome" title="Permalink to this definition"></a></dt>
<dd><p>Homer genome label to use.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.size">
<span class="sig-name descname"><span class="pre">size</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.size" title="Permalink to this definition"></a></dt>
<dd><p>Fragment size to use for motif finding. Default: ‘given’ [uses the exact regions you give it]</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.mask">
<span class="sig-name descname"><span class="pre">mask</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.mask" title="Permalink to this definition"></a></dt>
<dd><p>Whether to mask repeats or not. Default: True</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>bool, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.denovo">
<span class="sig-name descname"><span class="pre">denovo</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.denovo" title="Permalink to this definition"></a></dt>
<dd><p>Whether to infer overrepresented motifs de novo. Default: False</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>bool, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.length">
<span class="sig-name descname"><span class="pre">length</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.length" title="Permalink to this definition"></a></dt>
<dd><p>Motif length values. Default: 8,10,12</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.meme_path">
<span class="sig-name descname"><span class="pre">meme_path</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.meme_path" title="Permalink to this definition"></a></dt>
<dd><p>Path to meme bin folder. Meme will be used if given for motif annotation. Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.meme_collection_path">
<span class="sig-name descname"><span class="pre">meme_collection_path</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.meme_collection_path" title="Permalink to this definition"></a></dt>
<dd><p>Path to motif collection (in .cb format) to compare homer motifs with. Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.path_to_motif_annotations">
<span class="sig-name descname"><span class="pre">path_to_motif_annotations</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.path_to_motif_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Path to motif annotations. If not provided, they will be downloaded from
<a class="reference external" href="https://resources.aertslab.org">https://resources.aertslab.org</a> based on the specie name provided (only possible for mus_musculus,
homo_sapiens and drosophila_melanogaster). Default: None</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.annotation_version">
<span class="sig-name descname"><span class="pre">annotation_version</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.annotation_version" title="Permalink to this definition"></a></dt>
<dd><p>Motif collection version. Default: v9</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>str, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.cistrome_annotation">
<span class="sig-name descname"><span class="pre">cistrome_annotation</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.cistrome_annotation" title="Permalink to this definition"></a></dt>
<dd><p>Annotation to use for forming cistromes. It can be ‘Direct_annot’ (direct evidence that the motif is
linked to that TF), ‘Motif_similarity_annot’ (based on tomtom motif similarity), ‘Orthology_annot’
(based on orthology with a TF that is directly linked to that motif) or ‘Motif_similarity_and_Orthology_annot’.
Default: [‘Direct_annot’, ‘Motif_similarity_annot’, ‘Orthology_annot’, ‘Motif_similarity_and_Orthology_annot’]</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>List, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.motif_similarity_fdr">
<span class="sig-name descname"><span class="pre">motif_similarity_fdr</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.motif_similarity_fdr" title="Permalink to this definition"></a></dt>
<dd><p>Minimal motif similarity value to consider two motifs similar. Default: 0.001</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.orthologous_identity_threshold">
<span class="sig-name descname"><span class="pre">orthologous_identity_threshold</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.orthologous_identity_threshold" title="Permalink to this definition"></a></dt>
<dd><p>Minimal orthology value for considering two TFs orthologous. Default: 0.0</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>float, optional</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.known_motifs">
<span class="sig-name descname"><span class="pre">known_motifs</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.known_motifs" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe containing known motif enrichment results.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.denovo_motifs">
<span class="sig-name descname"><span class="pre">denovo_motifs</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.denovo_motifs" title="Permalink to this definition"></a></dt>
<dd><p>A dataframe containing de novo motif enrichment results.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.known_motif_hits">
<span class="sig-name descname"><span class="pre">known_motif_hits</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.known_motif_hits" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary containing regions with motif hits for each known motif.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.denovo_motif_hits">
<span class="sig-name descname"><span class="pre">denovo_motif_hits</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.denovo_motif_hits" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary containing regions with motif hits for each de novo motif.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.known_cistromes">
<span class="sig-name descname"><span class="pre">known_cistromes</span></span><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.known_cistromes" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary containing regions with motif hits for each TF found with known motifs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="id0">
<span class="sig-name descname"><span class="pre">denovo_motif_hits</span></span><a class="headerlink" href="#id0" title="Permalink to this definition"></a></dt>
<dd><p>A dictionary containing regions with motif hits for each TF found de novo.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>Dict</p>
</dd>
</dl>
</dd></dl>

<p class="rubric">References</p>
<p>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining
Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities.
Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</p>
<p class="rubric">Methods</p>
<table class="autosummary longtable docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 90%" />
</colgroup>
<tbody>
<tr class="row-odd"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_homer.Homer.add_motif_annotation_homer" title="pycistarget.motif_enrichment_homer.Homer.add_motif_annotation_homer"><code class="xref py py-obj docutils literal notranslate"><span class="pre">add_motif_annotation_homer</span></code></a>()</p></td>
<td><p>Add motif annotations (based on Homer, cisTarget and meme if specified)</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_homer.Homer.find_motif_hits" title="pycistarget.motif_enrichment_homer.Homer.find_motif_hits"><code class="xref py py-obj docutils literal notranslate"><span class="pre">find_motif_hits</span></code></a>([n_cpu])</p></td>
<td><p>Find motif hits with <cite>homer2 find</cite></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_homer.Homer.get_cistromes" title="pycistarget.motif_enrichment_homer.Homer.get_cistromes"><code class="xref py py-obj docutils literal notranslate"><span class="pre">get_cistromes</span></code></a>([annotation])</p></td>
<td><p>Format cistromes per TF</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_homer.Homer.load_denovo" title="pycistarget.motif_enrichment_homer.Homer.load_denovo"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_denovo</span></code></a>()</p></td>
<td><p>Load de novo motif enrichment results from file.</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_homer.Homer.load_known" title="pycistarget.motif_enrichment_homer.Homer.load_known"><code class="xref py py-obj docutils literal notranslate"><span class="pre">load_known</span></code></a>()</p></td>
<td><p>Load known motif enrichment results from file.</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference internal" href="#pycistarget.motif_enrichment_homer.Homer.run" title="pycistarget.motif_enrichment_homer.Homer.run"><code class="xref py py-obj docutils literal notranslate"><span class="pre">run</span></code></a>()</p></td>
<td><p>Run Homer</p></td>
</tr>
</tbody>
</table>
<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.add_motif_annotation_homer">
<span class="sig-name descname"><span class="pre">add_motif_annotation_homer</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#Homer.add_motif_annotation_homer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.add_motif_annotation_homer" title="Permalink to this definition"></a></dt>
<dd><p>Add motif annotations (based on Homer, cisTarget and meme if specified)</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.find_motif_hits">
<span class="sig-name descname"><span class="pre">find_motif_hits</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n_cpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#Homer.find_motif_hits"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.find_motif_hits" title="Permalink to this definition"></a></dt>
<dd><p>Find motif hits with <cite>homer2 find</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>n_cpu</strong> (<em>int</em>) – Number of cores to use.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.get_cistromes">
<span class="sig-name descname"><span class="pre">get_cistromes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#Homer.get_cistromes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.get_cistromes" title="Permalink to this definition"></a></dt>
<dd><p>Format cistromes per TF</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>cistrome_annotation</strong> (<em>List</em><em>, </em><em>optional</em>) – Annotation to use for forming cistromes. It can be ‘Direct_annot’ (direct evidence that the motif is
linked to that TF), ‘Motif_similarity_annot’ (based on tomtom motif similarity), ‘Orthology_annot’
(based on orthology with a TF that is directly linked to that motif) or ‘Motif_similarity_and_Orthology_annot’.
Default: [‘Direct_annot’, ‘Motif_similarity_annot’, ‘Orthology_annot’, ‘Motif_similarity_and_Orthology_annot’]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.load_denovo">
<span class="sig-name descname"><span class="pre">load_denovo</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#Homer.load_denovo"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.load_denovo" title="Permalink to this definition"></a></dt>
<dd><p>Load de novo motif enrichment results from file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.load_known">
<span class="sig-name descname"><span class="pre">load_known</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#Homer.load_known"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.load_known" title="Permalink to this definition"></a></dt>
<dd><p>Load known motif enrichment results from file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.Homer.run">
<span class="sig-name descname"><span class="pre">run</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#Homer.run"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.Homer.run" title="Permalink to this definition"></a></dt>
<dd><p>Run Homer</p>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.homer_results">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_homer.</span></span><span class="sig-name descname"><span class="pre">homer_results</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">homer_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">results</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'known'</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#homer_results"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.homer_results" title="Permalink to this definition"></a></dt>
<dd><p>A function to show Homer results in jupyter notebooks.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>Homer_dict</strong> (<em>Dict</em>) – A dictionary with one <a class="reference internal" href="#pycistarget.motif_enrichment_homer.Homer" title="pycistarget.motif_enrichment_homer.Homer"><code class="xref py py-class docutils literal notranslate"><span class="pre">Homer</span></code></a> object per slot.</p></li>
<li><p><strong>name</strong> (<em>str</em>) – Dictionary key of the analysis result to show. Default: None (All)</p></li>
<li><p><strong>results</strong> (<em>str</em>) – Whether to show know or de novo results. Default: ‘known’</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.motif_enrichment_homer.run_homer">
<span class="sig-prename descclassname"><span class="pre">pycistarget.motif_enrichment_homer.</span></span><span class="sig-name descname"><span class="pre">run_homer</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">homer_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outdir</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'given'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">denovo</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">bool</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">length</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'8,10,12'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cpu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">int</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meme_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meme_collection_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motif_annotations</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation_version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'v9'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cistrome_annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_similarity_fdr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orthologous_identity_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/motif_enrichment_homer.html#run_homer"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.motif_enrichment_homer.run_homer" title="Permalink to this definition"></a></dt>
<dd><p>Run Homer</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>homer_path</strong> (<em>str</em>) – Path to Homer bin folder.</p></li>
<li><p><strong>region_sets</strong> (<em>Dict</em>) – A dictionary of PyRanges containing region coordinates for the region sets to be analyzed.</p></li>
<li><p><strong>outdir</strong> (<em>str</em>) – Path to folder to output Homer results.</p></li>
<li><p><strong>genome</strong> (<em>str</em>) – Homer genome label to use.</p></li>
<li><p><strong>size</strong> (<em>str</em><em>, </em><em>optional</em>) – Fragment size to use for motif finding. Default: ‘given’ [uses the exact regions you give it]</p></li>
<li><p><strong>mask</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to mask repeats or not. Default: True</p></li>
<li><p><strong>denovo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to infer overrepresented motifs de novo. Default: False</p></li>
<li><p><strong>length</strong> (<em>str</em><em>, </em><em>optional</em>) – Motif length values. Default: 8,10,12</p></li>
<li><p><strong>n_cpu</strong> (<em>int</em>) – Number of cores to use.</p></li>
<li><p><strong>meme_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to meme bin folder. Meme will be used if given for motif annotation. Default: None</p></li>
<li><p><strong>meme_collection_path</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to motif collection (in .cb format) to compare homer motifs with. Default: None</p></li>
<li><p><strong>path_to_motif_annotations</strong> (<em>str</em><em>, </em><em>optional</em>) – Path to motif annotations. If not provided, they will be downloaded from
<a class="reference external" href="https://resources.aertslab.org">https://resources.aertslab.org</a> based on the specie name provided (only possible for mus_musculus,
homo_sapiens and drosophila_melanogaster). Default: None</p></li>
<li><p><strong>annotation_version</strong> (<em>str</em><em>, </em><em>optional</em>) – Motif collection version. Default: v9</p></li>
<li><p><strong>cistrome_annotation</strong> (<em>List</em><em>, </em><em>optional</em>) – Annotation to use for forming cistromes. It can be ‘Direct_annot’ (direct evidence that the motif is
linked to that TF), ‘Motif_similarity_annot’ (based on tomtom motif similarity), ‘Orthology_annot’
(based on orthology with a TF that is directly linked to that motif) or ‘Motif_similarity_and_Orthology_annot’.
Default: [‘Direct_annot’, ‘Motif_similarity_annot’, ‘Orthology_annot’, ‘Motif_similarity_and_Orthology_annot’]</p></li>
<li><p><strong>motif_similarity_fdr</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal motif similarity value to consider two motifs similar. Default: 0.001</p></li>
<li><p><strong>orthologous_identity_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimal orthology value for considering two TFs orthologous. Default: 0.0</p></li>
<li><p><strong>**kwargs</strong> – Extra parameters to pass to <cite>ray.init()</cite>.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining
Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities.
Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</p>
</dd></dl>

</div>
<div class="section" id="module-pycistarget.cluster_buster">
<span id="cluster-buster"></span><h2>Cluster-Buster<a class="headerlink" href="#module-pycistarget.cluster_buster" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.cluster_buster.cluster_buster">
<span class="sig-prename descclassname"><span class="pre">pycistarget.cluster_buster.</span></span><span class="sig-name descname"><span class="pre">cluster_buster</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">cbust_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_sets</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">PyRanges</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">Dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_genome_fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_regions_fasta</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_cpu</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motifs</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">bool</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/cluster_buster.html#cluster_buster"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.cluster_buster.cluster_buster" title="Permalink to this definition"></a></dt>
<dd><p>Add motif annotation</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>cluster_buster_path</strong> (<em>str</em>) – Path to cluster buster bin.</p></li>
<li><p><strong>path_to_motifs</strong> (<em>str</em><em>, </em><em>optional.</em>) – Path to motif collection folder (in .cb format). Only required if using a shuffled background.</p></li>
<li><p><strong>region_sets</strong> (<em>Dict</em>) – A dictionary of PyRanges containing region coordinates for the regions to be analyzed. Only required
if <cite>path_to_regions_fasta</cite> is not provided.</p></li>
<li><p><strong>path_to_genome_fasta</strong> (<em>str</em><em>, </em><em>optional.</em>) – Path to genome fasta file. Only required if <cite>path_to_regions_fasta</cite> is not provided. Default: None</p></li>
<li><p><strong>path_to_regions_fasta</strong> (<em>str</em><em>, </em><em>optional.</em>) – Path to regions fasta file. Only required if <cite>path_to_genome_fasta</cite> is not provided. Default: None</p></li>
<li><p><strong>n_cpu</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of cores to use</p></li>
<li><p><strong>motifs</strong> (<em>List</em><em>, </em><em>optional</em>) – Names of the motif files to use (from <cite>path_to_motifs</cite>). Default: None (All)</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to print progress to screen</p></li>
<li><p><strong>**kwargs</strong> – Additional parameters to pass to <cite>ray.init()</cite></p></li>
</ul>
</dd>
</dl>
<p class="rubric">References</p>
<p>Frith, Martin C., Michael C. Li, and Zhiping Weng. “Cluster-Buster: Finding dense clusters of motifs
in DNA sequences.” Nucleic acids research 31, no. 13 (2003): 3666-3668.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.cluster_buster.get_sequence_names_from_fasta">
<span class="sig-prename descclassname"><span class="pre">pycistarget.cluster_buster.</span></span><span class="sig-name descname"><span class="pre">get_sequence_names_from_fasta</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fasta_filename</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/cluster_buster.html#get_sequence_names_from_fasta"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.cluster_buster.get_sequence_names_from_fasta" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve sequence names from fasta</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.cluster_buster.grep">
<span class="sig-prename descclassname"><span class="pre">pycistarget.cluster_buster.</span></span><span class="sig-name descname"><span class="pre">grep</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">l</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">s</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/cluster_buster.html#grep"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.cluster_buster.grep" title="Permalink to this definition"></a></dt>
<dd><p>Helper for grep</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.cluster_buster.pyranges2names">
<span class="sig-prename descclassname"><span class="pre">pycistarget.cluster_buster.</span></span><span class="sig-name descname"><span class="pre">pyranges2names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PyRanges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/cluster_buster.html#pyranges2names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.cluster_buster.pyranges2names" title="Permalink to this definition"></a></dt>
<dd><p>Convert pyranges to sequence name (fasta format)</p>
</dd></dl>

</div>
<div class="section" id="module-pycistarget.utils">
<span id="utils"></span><h2>Utils<a class="headerlink" href="#module-pycistarget.utils" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.coord_to_region_names">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">coord_to_region_names</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">coord</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">PyRanges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#coord_to_region_names"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.coord_to_region_names" title="Permalink to this definition"></a></dt>
<dd><p>Convert coordinates to region names (UCSC format)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.get_TF_list">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">get_TF_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_enrichment_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#get_TF_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.get_TF_list" title="Permalink to this definition"></a></dt>
<dd><p>Get TFs from motif enrichment tables</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.get_cistrome_per_TF">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">get_cistrome_per_TF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_hits_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motifs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#get_cistrome_per_TF"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.get_cistrome_per_TF" title="Permalink to this definition"></a></dt>
<dd><p>Format cistromes per TF</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.get_cistromes_per_region_set">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">get_cistromes_per_region_set</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_enrichment_region_set</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_hits_regions_set</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#get_cistromes_per_region_set"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.get_cistromes_per_region_set" title="Permalink to this definition"></a></dt>
<dd><p>Get (direct/extended) cistromes for TFs</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.get_motifs_per_TF">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">get_motifs_per_TF</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">motif_enrichment_table</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">DataFrame</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tf</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_column</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">annotation</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">['Direct_annot',</span> <span class="pre">'Motif_similarity_annot',</span> <span class="pre">'Orthology_annot',</span> <span class="pre">'Motif_similarity_and_Orthology_annot']</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#get_motifs_per_TF"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.get_motifs_per_TF" title="Permalink to this definition"></a></dt>
<dd><p>Get motif annotated to each TF from a motif enrichment table</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.get_position_index">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">get_position_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#get_position_index"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.get_position_index" title="Permalink to this definition"></a></dt>
<dd><p>Get position of a query within a list</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.homer2meme">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">homer2meme</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">homer_motif_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#homer2meme"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.homer2meme" title="Permalink to this definition"></a></dt>
<dd><p>Convert Homer motifs to meme format</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.inplace_change">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">inplace_change</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filename</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">old_string</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">new_string</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#inplace_change"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.inplace_change" title="Permalink to this definition"></a></dt>
<dd><p>Replace string in a file</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.load_motif_annotations">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">load_motif_annotations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">specie</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">'v9'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Optional</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('#motif_id',</span> <span class="pre">'gene_name',</span> <span class="pre">'motif_similarity_qvalue',</span> <span class="pre">'orthologous_identity',</span> <span class="pre">'description')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">motif_similarity_fdr</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.001</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">orthologous_identity_threshold</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.0</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#load_motif_annotations"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.load_motif_annotations" title="Permalink to this definition"></a></dt>
<dd><p>Load motif annotations from a motif2TF snapshot.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>specie</strong> – Specie to retrieve annotations for.</p></li>
<li><p><strong>version</strong> – Motif collection version.</p></li>
<li><p><strong>fname</strong> – The snapshot taken from motif2TF.</p></li>
<li><p><strong>column_names</strong> – The names of the columns in the snapshot to load.</p></li>
<li><p><strong>motif_similarity_fdr</strong> – The maximum False Discovery Rate to find factor annotations for enriched motifs.</p></li>
<li><p><strong>orthologuous_identity_threshold</strong> – The minimum orthologuous identity to find factor annotations for enriched motifs.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.region_names_to_coordinates">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">region_names_to_coordinates</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region_names</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">List</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#region_names_to_coordinates"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.region_names_to_coordinates" title="Permalink to this definition"></a></dt>
<dd><p>Convert region names (UCSC format) to coordinates (pd.DataFrame)</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.region_sets_to_signature">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">region_sets_to_signature</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region_set</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">region_set_name</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#region_sets_to_signature"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.region_sets_to_signature" title="Permalink to this definition"></a></dt>
<dd><p>Generates a gene signature object from a dict of PyRanges objects</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pr_region_set</strong> – PyRanges object to be converted in genesignature object</p></li>
<li><p><strong>region_set_name</strong> – Name of the regions set</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.target_to_query">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">target_to_query</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">target</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">PyRanges</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">query</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">Union</span><span class="p"><span class="pre">[</span></span><span class="pre">PyRanges</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">List</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fraction_overlap</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">float</span></span><span class="w"> </span><span class="o"><span class="pre">=</span></span><span class="w"> </span><span class="default_value"><span class="pre">0.4</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#target_to_query"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.target_to_query" title="Permalink to this definition"></a></dt>
<dd><p>Map query regions to another set of regions</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="pycistarget.utils.tomtom">
<span class="sig-prename descclassname"><span class="pre">pycistarget.utils.</span></span><span class="sig-name descname"><span class="pre">tomtom</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">homer_motif_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meme_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">meme_collection_path</span></span><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="n"><span class="pre">str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/pycistarget/utils.html#tomtom"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#pycistarget.utils.tomtom" title="Permalink to this definition"></a></dt>
<dd><p>Run tomtom for Homer motif annotation</p>
</dd></dl>

</div>
</div>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="pycistarget_scenic%2B_human_brain.html" class="btn btn-neutral float-left" title="PycisTarget on human cerebellum DARs and topics" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Carmen Bravo Gonzalez Blas, Seppe de Winter.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>