<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Using pycisTarget within the SCENIC+ workflow &mdash; pycistarget 1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="_static/doctools.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> pycistarget
          </a>
              <div class="version">
                1.0.1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tools.html">pycisTarget methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="motif_table.html">SCENIC+ motif collection</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="api.html">API</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pycistarget</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Using pycisTarget within the SCENIC+ workflow</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/pycistarget_scenic+_wrapper.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS for nbsphinx extension */

/* remove conflicting styling from Sphinx themes */
div.nbinput.container div.prompt *,
div.nboutput.container div.prompt *,
div.nbinput.container div.input_area pre,
div.nboutput.container div.output_area pre,
div.nbinput.container div.input_area .highlight,
div.nboutput.container div.output_area .highlight {
    border: none;
    padding: 0;
    margin: 0;
    box-shadow: none;
}

div.nbinput.container > div[class*=highlight],
div.nboutput.container > div[class*=highlight] {
    margin: 0;
}

div.nbinput.container div.prompt *,
div.nboutput.container div.prompt * {
    background: none;
}

div.nboutput.container div.output_area .highlight,
div.nboutput.container div.output_area pre {
    background: unset;
}

div.nboutput.container div.output_area div.highlight {
    color: unset;  /* override Pygments text color */
}

/* avoid gaps between output lines */
div.nboutput.container div[class*=highlight] pre {
    line-height: normal;
}

/* input/output containers */
div.nbinput.container,
div.nboutput.container {
    display: -webkit-flex;
    display: flex;
    align-items: flex-start;
    margin: 0;
    width: 100%;
}
@media (max-width: 540px) {
    div.nbinput.container,
    div.nboutput.container {
        flex-direction: column;
    }
}

/* input container */
div.nbinput.container {
    padding-top: 5px;
}

/* last container */
div.nblast.container {
    padding-bottom: 5px;
}

/* input prompt */
div.nbinput.container div.prompt pre {
    color: #307FC1;
}

/* output prompt */
div.nboutput.container div.prompt pre {
    color: #BF5B3D;
}

/* all prompts */
div.nbinput.container div.prompt,
div.nboutput.container div.prompt {
    width: 4.5ex;
    padding-top: 5px;
    position: relative;
    user-select: none;
}

div.nbinput.container div.prompt > div,
div.nboutput.container div.prompt > div {
    position: absolute;
    right: 0;
    margin-right: 0.3ex;
}

@media (max-width: 540px) {
    div.nbinput.container div.prompt,
    div.nboutput.container div.prompt {
        width: unset;
        text-align: left;
        padding: 0.4em;
    }
    div.nboutput.container div.prompt.empty {
        padding: 0;
    }

    div.nbinput.container div.prompt > div,
    div.nboutput.container div.prompt > div {
        position: unset;
    }
}

/* disable scrollbars on prompts */
div.nbinput.container div.prompt pre,
div.nboutput.container div.prompt pre {
    overflow: hidden;
}

/* input/output area */
div.nbinput.container div.input_area,
div.nboutput.container div.output_area {
    -webkit-flex: 1;
    flex: 1;
    overflow: auto;
}
@media (max-width: 540px) {
    div.nbinput.container div.input_area,
    div.nboutput.container div.output_area {
        width: 100%;
    }
}

/* input area */
div.nbinput.container div.input_area {
    border: 1px solid #e0e0e0;
    border-radius: 2px;
    /*background: #f5f5f5;*/
}

/* override MathJax center alignment in output cells */
div.nboutput.container div[class*=MathJax] {
    text-align: left !important;
}

/* override sphinx.ext.imgmath center alignment in output cells */
div.nboutput.container div.math p {
    text-align: left;
}

/* standard error */
div.nboutput.container div.output_area.stderr {
    background: #fdd;
}

/* ANSI colors */
.ansi-black-fg { color: #3E424D; }
.ansi-black-bg { background-color: #3E424D; }
.ansi-black-intense-fg { color: #282C36; }
.ansi-black-intense-bg { background-color: #282C36; }
.ansi-red-fg { color: #E75C58; }
.ansi-red-bg { background-color: #E75C58; }
.ansi-red-intense-fg { color: #B22B31; }
.ansi-red-intense-bg { background-color: #B22B31; }
.ansi-green-fg { color: #00A250; }
.ansi-green-bg { background-color: #00A250; }
.ansi-green-intense-fg { color: #007427; }
.ansi-green-intense-bg { background-color: #007427; }
.ansi-yellow-fg { color: #DDB62B; }
.ansi-yellow-bg { background-color: #DDB62B; }
.ansi-yellow-intense-fg { color: #B27D12; }
.ansi-yellow-intense-bg { background-color: #B27D12; }
.ansi-blue-fg { color: #208FFB; }
.ansi-blue-bg { background-color: #208FFB; }
.ansi-blue-intense-fg { color: #0065CA; }
.ansi-blue-intense-bg { background-color: #0065CA; }
.ansi-magenta-fg { color: #D160C4; }
.ansi-magenta-bg { background-color: #D160C4; }
.ansi-magenta-intense-fg { color: #A03196; }
.ansi-magenta-intense-bg { background-color: #A03196; }
.ansi-cyan-fg { color: #60C6C8; }
.ansi-cyan-bg { background-color: #60C6C8; }
.ansi-cyan-intense-fg { color: #258F8F; }
.ansi-cyan-intense-bg { background-color: #258F8F; }
.ansi-white-fg { color: #C5C1B4; }
.ansi-white-bg { background-color: #C5C1B4; }
.ansi-white-intense-fg { color: #A1A6B2; }
.ansi-white-intense-bg { background-color: #A1A6B2; }

.ansi-default-inverse-fg { color: #FFFFFF; }
.ansi-default-inverse-bg { background-color: #000000; }

.ansi-bold { font-weight: bold; }
.ansi-underline { text-decoration: underline; }


div.nbinput.container div.input_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight] > pre,
div.nboutput.container div.output_area div[class*=highlight].math,
div.nboutput.container div.output_area.rendered_html,
div.nboutput.container div.output_area > div.output_javascript,
div.nboutput.container div.output_area:not(.rendered_html) > img{
    padding: 5px;
    margin: 0;
}

/* fix copybtn overflow problem in chromium (needed for 'sphinx_copybutton') */
div.nbinput.container div.input_area > div[class^='highlight'],
div.nboutput.container div.output_area > div[class^='highlight']{
    overflow-y: hidden;
}

/* hide copybtn icon on prompts (needed for 'sphinx_copybutton') */
.prompt .copybtn {
    display: none;
}

/* Some additional styling taken form the Jupyter notebook CSS */
.jp-RenderedHTMLCommon table,
div.rendered_html table {
  border: none;
  border-collapse: collapse;
  border-spacing: 0;
  color: black;
  font-size: 12px;
  table-layout: fixed;
}
.jp-RenderedHTMLCommon thead,
div.rendered_html thead {
  border-bottom: 1px solid black;
  vertical-align: bottom;
}
.jp-RenderedHTMLCommon tr,
.jp-RenderedHTMLCommon th,
.jp-RenderedHTMLCommon td,
div.rendered_html tr,
div.rendered_html th,
div.rendered_html td {
  text-align: right;
  vertical-align: middle;
  padding: 0.5em 0.5em;
  line-height: normal;
  white-space: normal;
  max-width: none;
  border: none;
}
.jp-RenderedHTMLCommon th,
div.rendered_html th {
  font-weight: bold;
}
.jp-RenderedHTMLCommon tbody tr:nth-child(odd),
div.rendered_html tbody tr:nth-child(odd) {
  background: #f5f5f5;
}
.jp-RenderedHTMLCommon tbody tr:hover,
div.rendered_html tbody tr:hover {
  background: rgba(66, 165, 245, 0.2);
}

/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="Using-pycisTarget-within-the-SCENIC+-workflow">
<h1>Using pycisTarget within the SCENIC+ workflow<a class="headerlink" href="#Using-pycisTarget-within-the-SCENIC+-workflow" title="Permalink to this heading"></a></h1>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="k">matplotlib</span> inline
<span class="kn">import</span> <span class="nn">pycistarget</span>
<span class="n">pycistarget</span><span class="o">.</span><span class="n">__version__</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&#39;1.0.1.dev29+g95fd9fb&#39;
</pre></div></div>
</div>
<p>As part of the SCENIC+ workflow we provide a wrapper to run pycistarget with the recommended settings. This approach will run cistarget and DEM, in the original region sets (binarized topics, DARs, …) and without promoters when indicated (to improve the signal of non-promoter motifs if region sets include a large proportion of promoters). For DEM, we use as background regions in other regions sets in the dictionary. This step can be run using a dataset specific database or a precomputed
database.</p>
<div class="section" id="1.-Using-a-dataset-specific-database">
<h2>1. Using a dataset specific database<a class="headerlink" href="#1.-Using-a-dataset-specific-database" title="Permalink to this heading"></a></h2>
<p>If the consensus regions in the dataset do not overlap with regions in the precomputed database (SCREEN regions for mouse and human, cisTarget regions for Drosophila), we recommend to generate a dataset specific database using the consensus peaks as regions. When using a precomputed database, dataset regions that do not overlap with the precomputed regions will be removed from the analysis, as they will not be found in cistromes.</p>
<p>As a first step, we need to generate the dataset specific database using the code available at <a class="reference external" href="https://github.com/aertslab/create_cisTarget_databases">https://github.com/aertslab/create_cisTarget_databases</a>. Below you can find the basic steps to do so:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-bash notranslate"><div class="highlight"><pre><span></span>%%bash
<span class="c1"># Paths and parameters</span>
<span class="nv">consensdir</span><span class="o">=</span><span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/data/MACS_ATAC/iterative/peak_filtering_norm&#39;</span>
<span class="nv">outdir</span><span class="o">=</span><span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/ctx_db&#39;</span>
<span class="nv">tag</span><span class="o">=</span><span class="s1">&#39;cluster_V10_DPCL&#39;</span>
<span class="nv">genomefa</span><span class="o">=</span><span class="s1">&#39;/staging/leuven/stg_00002/lcb/resources/human/hg38/hg38.fa&#39;</span>
<span class="nv">ncpu</span><span class="o">=</span><span class="m">36</span>
<span class="nv">cbdir</span><span class="o">=</span><span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/cluster_motif_collection_V10_no_desso_no_factorbook/cluster_buster/&#39;</span>
<span class="c1"># Create outdir</span>
<span class="k">if</span> <span class="o">[</span> ! -d <span class="nv">$outdir</span> <span class="o">]</span>
<span class="k">then</span>
    mkdir -p <span class="nv">$outdir</span>
<span class="k">fi</span>
<span class="c1">#### List of motifs</span>
<span class="nv">motif_list</span><span class="o">=</span><span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/cluster_motif_collection_V10_no_desso_no_factorbook/annotated_motifs/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv&#39;</span>
<span class="c1">#### Get fasta sequences</span>
<span class="nb">echo</span> <span class="s2">&quot;Extracting FASTA ...&quot;</span>
module load BEDTools
bedtools getfasta -fi <span class="nv">$genomefa</span> -bed <span class="nv">$consensdir</span>/combined_summits_final.bed &gt; <span class="nv">$consensdir</span>/consensus_regions.fa
<span class="nb">echo</span> <span class="s2">&quot;Done.&quot;</span>
<span class="c1">#### Create scores DB</span>
<span class="nb">echo</span> <span class="s2">&quot;Creating scores DB files ...&quot;</span>
<span class="c1">#### Activate environment</span>
conda_initialize /staging/leuven/stg_00002/lcb/ghuls/software/miniconda3/
conda activate create_cistarget_databases
<span class="c1">#### Set ${create_cistarget_databases_dir} to https://github.com/aertslab/create_cisTarget_databases</span>
<span class="nv">create_cistarget_databases_dir</span><span class="o">=</span><span class="s1">&#39;/staging/leuven/stg_00002/lcb/ghuls/software/create_cisTarget_databases&#39;</span>
<span class="c1">#### Score the motifs in 1 chunks; we will use the non-redundant db here</span>
<span class="k">for</span> current_part in <span class="o">{</span><span class="m">1</span>..1<span class="o">}</span> <span class="p">;</span> <span class="k">do</span>
     <span class="si">${</span><span class="nv">create_cistarget_databases_dir</span><span class="si">}</span>/create_cistarget_motif_databases.py <span class="se">\</span>
         -f <span class="nv">$consensdir</span>/consensus_regions.fa <span class="se">\</span>
         -M <span class="nv">$cbdir</span> <span class="se">\</span>
         -m <span class="nv">$motif_list</span> <span class="se">\</span>
         -p <span class="si">${</span><span class="nv">current_part</span><span class="si">}</span> <span class="m">1</span> <span class="se">\</span>
         -o <span class="nv">$outdir</span>/<span class="nv">$tag</span> <span class="se">\</span>
         -t <span class="m">35</span> <span class="se">\</span>
         -l
<span class="k">done</span>
<span class="nb">echo</span> <span class="s2">&quot;Done.&quot;</span>
<span class="c1">#### Create rankings</span>
<span class="nb">echo</span> <span class="s2">&quot;Creating rankings DB files ...&quot;</span>
<span class="si">${</span><span class="nv">create_cistarget_databases_dir</span><span class="si">}</span>/convert_motifs_or_tracks_vs_regions_or_genes_scores_to_rankings_cistarget_dbs.py -i <span class="nv">$outdir</span>/<span class="nv">$tag</span>.motifs_vs_regions.scores.feather -s <span class="m">555</span>
<span class="nb">echo</span> <span class="s2">&quot;Done.&quot;</span>
<span class="nb">echo</span> <span class="s2">&quot;ALL DONE.&quot;</span>
</pre></div>
</div>
</div>
<p>Next you will need to load the region sets you want to include in the analysis. By default, we use binarized topics and DARs, but you can include additional analyses:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load region binarized topics</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="n">outDir</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/pycisTopic/&#39;</span>
<span class="n">infile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outDir</span><span class="o">+</span><span class="s1">&#39;topic_binarization/binarized_topic_region.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="n">binarized_topic_region</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">infile</span><span class="p">)</span>
<span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="c1"># Load DARs</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="n">infile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outDir</span><span class="o">+</span><span class="s1">&#39;DARs/DARs.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="n">DARs_dict</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">infile</span><span class="p">)</span>
<span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="c1"># Format region sets</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">pyranges</span> <span class="k">as</span> <span class="nn">pr</span>
<span class="kn">from</span> <span class="nn">pycistarget.utils</span> <span class="k">import</span> <span class="o">*</span>
<span class="n">region_sets</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">region_sets</span><span class="p">[</span><span class="s1">&#39;Topics&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">region_names_to_coordinates</span><span class="p">(</span><span class="n">binarized_topic_region</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">binarized_topic_region</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
<span class="n">region_sets</span><span class="p">[</span><span class="s1">&#39;DARs&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;[^A-Za-z0-9]+&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">region_names_to_coordinates</span><span class="p">(</span><span class="n">DARs_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">DARs_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
</pre></div>
</div>
</div>
<p>Finally we run the wrapper function. This will create a menr.pkl file in the output directory that can be directly used as input for SCENIC+.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run pycistarget</span>
<span class="c1"># run_without_promoters = True, will run the methods in all regions + the region sets without promoters</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/scenicplus/src/&#39;</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">scenicplus.wrappers.run_pycistarget</span> <span class="k">import</span> <span class="o">*</span>
<span class="n">run_pycistarget</span><span class="p">(</span><span class="n">region_sets</span><span class="p">,</span>
                 <span class="n">ctx_db_path</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/ctx_db/cluster_V10_DPCL.regions_vs_motifs.rankings.feather&#39;</span><span class="p">,</span>
                 <span class="n">species</span> <span class="o">=</span> <span class="s1">&#39;homo_sapiens&#39;</span><span class="p">,</span>
                 <span class="n">save_path</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/pycistarget/cluster_V10_V2/&#39;</span><span class="p">,</span>
                 <span class="n">dem_db_path</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/ctx_db/cluster_V10_DPCL.regions_vs_motifs.scores.feather&#39;</span><span class="p">,</span>
                 <span class="n">run_without_promoters</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                 <span class="n">biomart_host</span> <span class="o">=</span> <span class="s1">&#39;http://www.ensembl.org&#39;</span><span class="p">,</span>
                 <span class="n">promoter_space</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
                 <span class="n">ctx_auc_threshold</span> <span class="o">=</span> <span class="mf">0.005</span><span class="p">,</span>
                 <span class="n">ctx_nes_threshold</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
                 <span class="n">ctx_rank_threshold</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
                 <span class="n">dem_log2fc_thr</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
                 <span class="n">dem_motif_hit_thr</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
                 <span class="n">dem_max_bg_regions</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
                 <span class="n">path_to_motif_annotations</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/cluster_motif_collection_V10_no_desso_no_factorbook/snapshots/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv&#39;</span><span class="p">,</span>
                 <span class="n">annotation_version</span> <span class="o">=</span> <span class="s1">&#39;v10nr_clust&#39;</span><span class="p">,</span>
                 <span class="n">annotation</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Direct_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Orthology_annot&#39;</span><span class="p">],</span>
                 <span class="n">n_cpu</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                 <span class="n">_temp_dir</span> <span class="o">=</span> <span class="s1">&#39;/scratch/leuven/313/vsc31305/ray_spill&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="2.-Using-a-precomputed-database">
<h2>2. Using a precomputed database<a class="headerlink" href="#2.-Using-a-precomputed-database" title="Permalink to this heading"></a></h2>
<p>Alternatively you can use a precomputed database. This will reduce the running time as you can skip the generation of a database; however, regions not overlapping regions in the database will be missed. If the overlap is significant (&gt;80%), using this database will give similar results. Precomputed databases are available at <a class="reference external" href="https://resources.aertslab.org/cistarget/">https://resources.aertslab.org/cistarget/</a>.</p>
<p>We first load the region sets as before:</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load region binarized topics</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="n">outDir</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/pycisTopic/&#39;</span>
<span class="n">infile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outDir</span><span class="o">+</span><span class="s1">&#39;topic_binarization/binarized_topic_region.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="n">binarized_topic_region</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">infile</span><span class="p">)</span>
<span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="c1"># Load DARs</span>
<span class="kn">import</span> <span class="nn">pickle</span>
<span class="n">infile</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">outDir</span><span class="o">+</span><span class="s1">&#39;DARs/DARs.pkl&#39;</span><span class="p">,</span> <span class="s1">&#39;rb&#39;</span><span class="p">)</span>
<span class="n">DARs_dict</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">infile</span><span class="p">)</span>
<span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
<span class="c1"># Format region sets</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">import</span> <span class="nn">pyranges</span> <span class="k">as</span> <span class="nn">pr</span>
<span class="kn">from</span> <span class="nn">pycistarget.utils</span> <span class="k">import</span> <span class="o">*</span>
<span class="n">region_sets</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">region_sets</span><span class="p">[</span><span class="s1">&#39;Topics&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">region_names_to_coordinates</span><span class="p">(</span><span class="n">binarized_topic_region</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">binarized_topic_region</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
<span class="n">region_sets</span><span class="p">[</span><span class="s1">&#39;DARs&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="s1">&#39;[^A-Za-z0-9]+&#39;</span><span class="p">,</span> <span class="s1">&#39;_&#39;</span><span class="p">,</span> <span class="n">key</span><span class="p">):</span> <span class="n">pr</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">region_names_to_coordinates</span><span class="p">(</span><span class="n">DARs_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()))</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">DARs_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>
</pre></div>
</div>
</div>
<p>Finally, we run the pycistarget wrapper, but using the precomputed database instead.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Run pycistarget</span>
<span class="c1"># run_without_promoters = True, will run the methods in all regions + the region sets without promoters</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="n">os</span><span class="o">.</span><span class="n">chdir</span><span class="p">(</span><span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/scenicplus/src/&#39;</span><span class="p">)</span>
<span class="kn">from</span> <span class="nn">scenicplus.wrappers.run_pycistarget</span> <span class="k">import</span> <span class="o">*</span>
<span class="n">run_pycistarget</span><span class="p">(</span><span class="n">region_sets</span><span class="p">,</span>
                 <span class="n">ctx_db_path</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/hg38_SCREEN_cluster_db/cluster_SCREEN.regions_vs_motifs.rankings.feather&#39;</span><span class="p">,</span>
                 <span class="n">species</span> <span class="o">=</span> <span class="s1">&#39;homo_sapiens&#39;</span><span class="p">,</span>
                 <span class="n">save_path</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/DPCL/pycistarget/SCREEN_cluster_V10_V2/&#39;</span><span class="p">,</span>
                 <span class="n">dem_db_path</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/Multiomics_pipeline/analysis/hg38_SCREEN_cluster_db/cluster_SCREEN.regions_vs_motifs.scores.feather&#39;</span><span class="p">,</span>
                 <span class="n">run_without_promoters</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
                 <span class="n">biomart_host</span> <span class="o">=</span> <span class="s1">&#39;http://www.ensembl.org&#39;</span><span class="p">,</span>
                 <span class="n">promoter_space</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
                 <span class="n">ctx_auc_threshold</span> <span class="o">=</span> <span class="mf">0.005</span><span class="p">,</span>
                 <span class="n">ctx_nes_threshold</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
                 <span class="n">ctx_rank_threshold</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">,</span>
                 <span class="n">dem_log2fc_thr</span> <span class="o">=</span> <span class="mf">0.5</span><span class="p">,</span>
                 <span class="n">dem_motif_hit_thr</span> <span class="o">=</span> <span class="mf">3.0</span><span class="p">,</span>
                 <span class="n">dem_max_bg_regions</span> <span class="o">=</span> <span class="mi">500</span><span class="p">,</span>
                 <span class="n">path_to_motif_annotations</span> <span class="o">=</span> <span class="s1">&#39;/staging/leuven/stg_00002/lcb/cbravo/cluster_motif_collection_V10_no_desso_no_factorbook/snapshots/motifs-v10-nr.more_orthology.hgnc-mm0.00001-o0.0_clust.tsv&#39;</span><span class="p">,</span>
                 <span class="n">annotation_version</span> <span class="o">=</span> <span class="s1">&#39;v10nr_clust&#39;</span><span class="p">,</span>
                 <span class="n">annotation</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Direct_annot&#39;</span><span class="p">,</span> <span class="s1">&#39;Orthology_annot&#39;</span><span class="p">],</span>
                 <span class="n">n_cpu</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span>
                 <span class="n">_temp_dir</span> <span class="o">=</span> <span class="s1">&#39;/scratch/leuven/313/vsc31305/ray_spill&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Carmen Bravo Gonzalez Blas, Seppe de Winter.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>